<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Week 1: Importing Genetic Data • LandGenCourse</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Week 1: Importing Genetic Data">
<meta property="og:description" content="LandGenCourse">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-168837293-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-168837293-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LandGenCourse</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Week0_BasicR.html">Week 0: Intro to R Programming</a>
    </li>
    <li>
      <a href="../articles/Week0_Graphics.html">Week 0: R Graphics</a>
    </li>
    <li>
      <a href="../articles/Week1_vignette.html">Week 1: Importing Genetic Data</a>
    </li>
    <li>
      <a href="../articles/Week2_vignette.html">Week 2: Spatial Data</a>
    </li>
    <li>
      <a href="../articles/Week2_bonus_vignette.html">Week 2: Bonus Material</a>
    </li>
    <li>
      <a href="../articles/Week3_vignette.html">Week 3: Genetic Diversity</a>
    </li>
    <li>
      <a href="../articles/Week4_vignette.html">Week 4: Metapopulation Genetics</a>
    </li>
    <li>
      <a href="../articles/Week5_vignette.html">Week 5: Spatial Statistics</a>
    </li>
    <li>
      <a href="../articles/Week6_vignette.html">Week 6: Quantitative Genetics</a>
    </li>
    <li>
      <a href="../articles/Week7_vignette.html">Week 7: Spatial Linear Models</a>
    </li>
    <li>
      <a href="../articles/Week8_vignette.html">Week 8: Simulation</a>
    </li>
    <li>
      <a href="../articles/Week8_bonus_vignette.html">Week 8: Bonus Material</a>
    </li>
    <li>
      <a href="../articles/Week9_vignette.html">Week 9: Population Structure</a>
    </li>
    <li>
      <a href="../articles/Week10_vignette.html">Week 10: Landscape Resistance</a>
    </li>
    <li>
      <a href="../articles/Week11_vignette.html">Week 11: Detecting Adaptation</a>
    </li>
    <li>
      <a href="../articles/Week12_vignette.html">Week 12: Model Selection</a>
    </li>
    <li>
      <a href="../articles/Week13_vignette.html">Week 13: Gravity Models</a>
    </li>
    <li>
      <a href="../articles/Week14_vignette.html">Week 14: Contemporary Gene Flow</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../tutorials/00-Intro_RStudio.html">Introduction to RStudio</a>
    </li>
    <li>
      <a href="../tutorials/00-Intro_LandGenCourse.html">Introduction to the LandGenCourse package</a>
    </li>
    <li>
      <a href="../tutorials/00-Intro_RNotebooks.html">Introduction to R Notebooks</a>
    </li>
    <li>
      <a href="../tutorials/00-Version_Control.html">Version Control 101</a>
    </li>
    <li>
      <a href="../tutorials/01-Genetic_Data.html">Week 1 Getting Landscape Genetic Data into R</a>
    </li>
    <li>
      <a href="../tutorials/02-Spatial_Data.html">Week 2 Land Cover Analysis in R</a>
    </li>
    <li>
      <a href="../tutorials/03-Genetic_Diversity.html">Week 3 Genetic Diversity (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/03-Genetic_Diversity_Part2.html">Week 3 Genetic Diversity (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/04-Metapop_Genetics.html">Week 4 Metapopulation Genetics (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/04-Metapop_Genetics_Part2.html">Week 4 Metapopulation Genetics (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/05-Spatial_Stats.html">Week 5 Spatial Statistics</a>
    </li>
    <li>
      <a href="../tutorials/06-Quantitative_Genetics.html">Week 6 Quantitative Genetics (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/06-Quantitative_Genetics_Part2.html">Week 6 Quantitative Genetics (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/07-Spatial_Models.html">Week 7 Spatial Linear Models (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/07-Spatial_Models_Part2.html">Week 7 Spatial Linear Models (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/08-Modeling.html">Week 8 Simulation Experiments in R (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/08-Modeling_Part2.html">Week 8 Simulation Experiments in R (Part 2)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Week 1: Importing Genetic Data</h1>
                        <h4 class="author">Helene Wagner</h4>
            
            <h4 class="date">2020-06-08</h4>
      
      
      <div class="hidden name"><code>Week1_vignette.Rmd</code></div>

    </div>

    
    
<div id="overview-of-worked-example" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-of-worked-example" class="anchor"></a>1. Overview of Worked Example</h2>
<div id="a-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#a-goals" class="anchor"></a>a) Goals</h3>
<p>This worked example shows:</p>
<ul>
<li>How microsatellite data may be coded in a spreadsheet.</li>
<li>How such a file is imported into R.</li>
<li>How the data are imported into a ‘genind’ object using package ‘adegenet’.</li>
<li>How to view information stored in a ‘genind’ object.</li>
<li>How to import the data with the ‘gstudio’ package.</li>
</ul>
<p>Try modifying the code to import your own data!</p>
</div>
<div id="b-data-set" class="section level3">
<h3 class="hasAnchor">
<a href="#b-data-set" class="anchor"></a>b) Data set</h3>
<p>This code imports genetic data from 181 individuals of Colombia spotted frogs (Rana luteiventris) from 12 populations. The data are a subsample of the full data set analyzed in Funk et al. (2005) and Murphy et al. (2010). Please see the separate introduction to the data set.</p>
<ul>
<li>
<strong>ralu.loci</strong>: Data frame with populations and genetic data (181 rows x 9 columns). Included in package <code>LandGenCourse</code>.</li>
</ul>
</div>
<div id="c-required-r-libraries" class="section level3">
<h3 class="hasAnchor">
<a href="#c-required-r-libraries" class="anchor"></a>c) Required R libraries</h3>
<p>All required packages should have been installed already when you installed ‘LandGenCourse’.</p>
<p>Note: the function <code>library</code> will always load the package, even if it is already loaded, whereas <code>require</code> will only load it if it is not yet loaded. Either will work.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">adegenet</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">gstudio</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">LandGenCourse</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tibble</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">here</span>)</pre></body></html></div>
</div>
</div>
<div id="import-data-from--csv-file" class="section level2">
<h2 class="hasAnchor">
<a href="#import-data-from--csv-file" class="anchor"></a>2. Import data from .csv file</h2>
<div id="a-preparation" class="section level3">
<h3 class="hasAnchor">
<a href="#a-preparation" class="anchor"></a>a) Preparation</h3>
<p>Create a new R project for this lab: File &gt; New Project &gt; New Directory &gt; New Project.</p>
<p>This will automatically set your working directory to the new project folder. That’s where R will look for files and save anything you export.</p>
<p>Note: R does not distinguish between single quotes and double quotes, they are treated as synonyms, but the opening and closing symbols need to match.</p>
<p>The dataset <code>ralu.loci</code> is available in the package ‘LandGenCourse’ and can be loaded with <code><a href="https://rdrr.io/r/utils/data.html">data(ralu.loci)</a></code>.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">ralu.loci</span>)</pre></body></html></div>
<p>Below, however, we’ll import it from a comma separated file (.csv), as this is the recommended format for importing your own data (e.g. from Excel).</p>
</div>
<div id="b-excel-file" class="section level3">
<h3 class="hasAnchor">
<a href="#b-excel-file" class="anchor"></a>b) Excel file</h3>
<p>First we need to copy the file <code>ralu.loci.csv</code> to the <code>downloads</code> folder inside your project folder. The first line checks that the folder <code>downloads</code> exists and creates it if needed.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="kw">if</span>(!<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.exists</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span>(),<span class="st">"/downloads"</span>))) <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span>(),<span class="st">"/downloads"</span>))
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"ralu.loci.csv"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"LandGenCourse"</span>),
          <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span>(), <span class="st">"/downloads/ralu.loci.csv"</span>), <span class="kw">overwrite</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
<pre><code>## [1] FALSE</code></pre>
<p>Check that the file <code>ralu.loci.csv</code> is now in the folder <code>downloads</code> within your project folder. From your file system, you can open it with Excel. Note that in Excel, the columns with the loci must be explicitly defined as ‘text’, otherwise Excel is likely to interpret them as times.</p>
<p><strong>Excel instructions:</strong></p>
<p>Opening an existing .csv file with genetic data:</p>
<ul>
<li>Excel menu: ‘File &gt; New Workbook’.</li>
<li>Excel menu: ‘File &gt; Import &gt; CSV File &gt; Import’. Navigate to file.</li>
<li>Text Import Wizard: select ‘Delimited’, click ‘Next’.</li>
<li>Delimiters: check ‘comma’. Click ‘Next’.</li>
<li>Use the ‘Shift’ key to select all columns with loci A - H. Select ‘text’. Click ‘Finish &gt; OK’.</li>
</ul>
<p>Creating a new Excel file to enter genetic data:</p>
<ul>
<li>Label the loci columns.</li>
<li>Select all loci columns by clicking on the column headers (hold the ‘Shift’ key to select multiple adjacent columns).</li>
<li>Define the data format: ‘Format &gt; Cell &gt; Text’.</li>
<li>Enter genetic data, separating the alleles by a colon (e.g., ‘1:3’).</li>
</ul>
<p>Saving Excel file for import in R:</p>
<ul>
<li>File Format: ‘Comma Separated Values (.csv)’</li>
</ul>
<p>Here is a screenshot of the first few lines of the Excel file with the genetic data:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html">include_graphics</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"ExcelTable.png"</span>,
                            <span class="kw">package</span> <span class="kw">=</span> <span class="st">"LandGenCourse"</span>))</pre></body></html></div>
<p><img src="../../../../../../private/var/folders/kz/fv8pzy0d3cl_mp3md294cvbh0000gn/T/RtmpwBqDUn/temp_libpath2c3f20c3e939/LandGenCourse/extdata/ExcelTable.png" width="50%"></p>
</div>
<div id="c-import-with-function-read-csv" class="section level3">
<h3 class="hasAnchor">
<a href="#c-import-with-function-read-csv" class="anchor"></a>c) Import with function ‘read.csv’</h3>
<p>As the file is saved in csv format, we use the function ‘read.csv’ to import it into an R object called ‘Frogs’. Display the first few rows and columns to check whether the data have been imported correctly:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">Frogs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span>(), <span class="st">"/downloads/ralu.loci.csv"</span>), <span class="kw">header</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span>(<span class="no">Frogs</span>)</pre></body></html></div>
<pre><code>## # A tibble: 181 x 10
##    SiteName     Pop      A     B     C     D     E     F     G     H    
##    &lt;fct&gt;        &lt;fct&gt;    &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;
##  1 AirplaneLake Airplane 1:1   1:1   1:1   1:1   1:2   1:1   1:1   4:5  
##  2 AirplaneLake Airplane 2:2   1:1   NA:NA 1:1   1:1   NA:NA 2:2   NA:NA
##  3 AirplaneLake Airplane 1:1   1:1   1:1   1:1   3:3   1:1   1:1   3:3  
##  4 AirplaneLake Airplane 1:1   1:1   NA:NA 2:2   1:2   NA:NA NA:NA NA:NA
##  5 AirplaneLake Airplane 1:2   1:3   1:1   1:1   1:2   1:1   1:2   4:5  
##  6 AirplaneLake Airplane 1:2   1:1   1:1   3:1   1:1   1:1   1:2   4:5  
##  7 AirplaneLake Airplane 2:2   1:3   1:1   1:1   3:3   1:1   1:1   2:3  
##  8 AirplaneLake Airplane 2:2   1:3   1:1   1:1   3:3   1:1   1:1   2:3  
##  9 AirplaneLake Airplane 3:1   1:1   1:1   1:1   1:7   1:1   1:1   3:5  
## 10 AirplaneLake Airplane 2:2   1:3   1:1   1:1   3:7   1:1   1:1   3:3  
## # … with 171 more rows</code></pre>
<ul>
<li>The data frame ‘Frogs’ contains 181 rows and 11 columns. Only the first few are shown.</li>
<li>The file has a header row with column names.</li>
<li>Notice the abbreviations for the R data type under each variable name: <code>&lt;fctr&gt;</code> means that the variable has been coded as a factor.</li>
<li>Each row is one individual. Note that for this data set, there is no column with ID or names of individuals.</li>
<li>The first two columns indicates the site in long (“SiteName”) and short (“Pop”) format.</li>
<li>There are 8 columns with loci, named A - H. Each locus has two alleles, coded with numbers and separated by a colon (:).</li>
<li>This coding suggests that the markers are codominant and that the species is diploid.</li>
<li>The code for missing values is <code>NA</code>, (which shows up as <code>NA:NA</code> for missing genetic data). There seem to be quite a few missing values.</li>
</ul>
</div>
<div id="d-create-id-variable" class="section level3">
<h3 class="hasAnchor">
<a href="#d-create-id-variable" class="anchor"></a>d) Create ID variable</h3>
<p>We’ll add unique ID’s for the frogs (to show how these would be imported). Here we take, for each frog, a substring of <code>Pop</code> containing the first three letters, add the row number, and separate them by a period.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">Frogs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">FrogID</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span>(<span class="no">Frogs</span>$<span class="no">Pop</span>, <span class="fl">1</span>, <span class="fl">3</span>), <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span>(<span class="no">Frogs</span>), <span class="kw">sep</span><span class="kw">=</span><span class="st">"."</span>), <span class="no">Frogs</span>)
<span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span>(<span class="no">Frogs</span>)</pre></body></html></div>
<pre><code>## # A tibble: 181 x 11
##    FrogID SiteName     Pop      A     B     C     D     E     F     G     H    
##    &lt;fct&gt;  &lt;fct&gt;        &lt;fct&gt;    &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;
##  1 Air.1  AirplaneLake Airplane 1:1   1:1   1:1   1:1   1:2   1:1   1:1   4:5  
##  2 Air.2  AirplaneLake Airplane 2:2   1:1   NA:NA 1:1   1:1   NA:NA 2:2   NA:NA
##  3 Air.3  AirplaneLake Airplane 1:1   1:1   1:1   1:1   3:3   1:1   1:1   3:3  
##  4 Air.4  AirplaneLake Airplane 1:1   1:1   NA:NA 2:2   1:2   NA:NA NA:NA NA:NA
##  5 Air.5  AirplaneLake Airplane 1:2   1:3   1:1   1:1   1:2   1:1   1:2   4:5  
##  6 Air.6  AirplaneLake Airplane 1:2   1:1   1:1   3:1   1:1   1:1   1:2   4:5  
##  7 Air.7  AirplaneLake Airplane 2:2   1:3   1:1   1:1   3:3   1:1   1:1   2:3  
##  8 Air.8  AirplaneLake Airplane 2:2   1:3   1:1   1:1   3:3   1:1   1:1   2:3  
##  9 Air.9  AirplaneLake Airplane 3:1   1:1   1:1   1:1   1:7   1:1   1:1   3:5  
## 10 Air.10 AirplaneLake Airplane 2:2   1:3   1:1   1:1   3:7   1:1   1:1   3:3  
## # … with 171 more rows</code></pre>
</div>
<div id="e-export-with-function-write-csv" class="section level3">
<h3 class="hasAnchor">
<a href="#e-export-with-function-write-csv" class="anchor"></a>e) Export with function ‘write.csv’</h3>
<p>If you want to export (and import) your own files, adapt this code.</p>
<p>Note: when using R Notebooks, problems can arise if R looks for files in a different place (more on this in Week 8 Bonus material). This is a safe work-around that will help make your work reproducible and transferable to other computers. For this to work properly, you must work within an R project, and you must have writing permission for the project folder.</p>
<p>The function <code>here</code> from the package <code>here</code> finds the path to your project folder. Remove the hashtags and run the two lines to see the path to your project folder, and the path to the (to be created) output folder. The function <code>paste0</code> simply pastes strings of text together, without any separator (hence the ‘0’ in the function name).</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r">#here() 
#paste0(here(),"/output")</pre></body></html></div>
<p>The following line checks whether an <code>output</code> folder exists in the project folder, and if not, creates one.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="kw">if</span>(!<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.exists</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span>(),<span class="st">"/output"</span>))) <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span>(),<span class="st">"/output"</span>))</pre></body></html></div>
<p>Now we can write our data into a csv file in the output folder:</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.csv</a></span>(<span class="no">ralu.loci</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span>(),<span class="st">"/output/ralu.loci.csv"</span>),
          <span class="kw">quote</span><span class="kw">=</span><span class="fl">FALSE</span>, <span class="kw">row.names</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
<p>To import the file again from the output folder, you could do this:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">ralu.loci.2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span>(),<span class="st">"/output/ralu.loci.csv"</span>), <span class="kw">header</span><span class="kw">=</span><span class="fl">TRUE</span>)</pre></body></html></div>
</div>
</div>
<div id="create-genind-object" class="section level2">
<h2 class="hasAnchor">
<a href="#create-genind-object" class="anchor"></a>3. Create ‘genind’ Object</h2>
<div id="a-read-the-helpfile" class="section level3">
<h3 class="hasAnchor">
<a href="#a-read-the-helpfile" class="anchor"></a>a) Read the helpfile</h3>
<p>We use the function <code>df2genind</code> of the <code>adegenet</code> package to import the loci into a <code>genind</code> object named <code>Frogs.genind</code>. Check the help file for a definition of all arguments and for example code.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r">?<span class="no">df2genind</span></pre></body></html></div>
</div>
<div id="b-set-all-arguments-of-function-df2genind" class="section level3">
<h3 class="hasAnchor">
<a href="#b-set-all-arguments-of-function-df2genind" class="anchor"></a>b) Set all arguments of function ‘df2genind’</h3>
<p>Some explanations:</p>
<ul>
<li>
<strong>X</strong>: this is the data frame (or matrix) containing the loci only. Hence we need columns 4 - 11.</li>
<li>
<strong>sep</strong>: need to specify the separator between alleles (here: colon)</li>
<li>
<strong>ncode</strong>: not needed here because we used a separator, defined under <code>sep</code>.</li>
<li>
<strong>ind.names</strong>: here we should indicate the column <code>FrogID</code>.</li>
<li>
<strong>loc.names</strong>: by default this will use the column names of the loci.</li>
<li>
<strong>pop</strong>: here we should indicate either the column <code>Pop</code> or <code>SiteName</code>.</li>
<li>
<strong>NA.char</strong>: how were missing values coded? Here: <code>NA</code>.</li>
<li>
<strong>ploidy</strong>: this species is diploid, hence <code>2</code>.</li>
<li>
<strong>type</strong>: marker type, either <code>"codom"</code> for codominant markers like microsatellites, or <code>"PA"</code> for presence-absence, such as SNP or AFLP markers. Note: you can use either "" or ’’.</li>
<li>
<strong>strata</strong>: one way of defining hierarchical sampling levels. We will add this in Week 3.</li>
<li>
<strong>hierarchy</strong>: another way of defining hierarchical sampling levels. We’ll skip it for now.</li>
</ul>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">Frogs.genind</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/df2genind.html">df2genind</a></span>(<span class="kw">X</span><span class="kw">=</span><span class="no">Frogs</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">4</span>:<span class="fl">11</span>)], <span class="kw">sep</span><span class="kw">=</span><span class="st">":"</span>, <span class="kw">ncode</span><span class="kw">=</span><span class="kw">NULL</span>, <span class="kw">ind.names</span><span class="kw">=</span> <span class="no">Frogs</span>$<span class="no">FrogID</span>, <span class="kw">loc.names</span><span class="kw">=</span><span class="kw">NULL</span>, <span class="kw">pop</span><span class="kw">=</span><span class="no">Frogs</span>$<span class="no">Pop</span>, <span class="kw">NA.char</span><span class="kw">=</span><span class="st">"NA"</span>, <span class="kw">ploidy</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"codom"</span>, <span class="kw">strata</span><span class="kw">=</span><span class="kw">NULL</span>, <span class="kw">hierarchy</span><span class="kw">=</span><span class="kw">NULL</span>)</pre></body></html></div>
</div>
<div id="c-check-imported-data" class="section level3">
<h3 class="hasAnchor">
<a href="#c-check-imported-data" class="anchor"></a>c) Check imported data</h3>
<p>Printing the genind object’s name lists the number of individuals, loci and alleles, and lists all slots (prefaced by @).</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">Frogs.genind</span></pre></body></html></div>
<pre><code>## /// GENIND OBJECT /////////
## 
##  // 181 individuals; 8 loci; 39 alleles; size: 55.2 Kb
## 
##  // Basic content
##    @tab:  181 x 39 matrix of allele counts
##    @loc.n.all: number of alleles per locus (range: 3-9)
##    @loc.fac: locus factor for the 39 columns of @tab
##    @all.names: list of allele names for each locus
##    @ploidy: ploidy of each individual  (range: 2-2)
##    @type:  codom
##    @call: df2genind(X = Frogs[, c(4:11)], sep = ":", ncode = NULL, ind.names = Frogs$FrogID, 
##     loc.names = NULL, pop = Frogs$Pop, NA.char = "NA", ploidy = 2, 
##     type = "codom", strata = NULL, hierarchy = NULL)
## 
##  // Optional content
##    @pop: population of each individual (group size range: 7-23)</code></pre>
</div>
<div id="d-summarize-genind-object" class="section level3">
<h3 class="hasAnchor">
<a href="#d-summarize-genind-object" class="anchor"></a>d) Summarize ‘genind’ object</h3>
<p>There is a <code>summary</code> function for <code>genind</code> objects. Notes:</p>
<ul>
<li>
<code>Group</code> here refers to the populations, i.e., <code>Group sizes</code> means number of individuals per population.</li>
<li>It is a good idea to check the percentage of missing values here. If it is 0%, this may indicate that the coding for missing values was not recognized. If the % missing is very high, on the other hand, there may be some other importing problem.</li>
</ul>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">Frogs.genind</span>)</pre></body></html></div>
<pre><code>## 
## // Number of individuals: 181
## // Group sizes: 21 8 14 13 7 17 9 20 19 13 17 23
## // Number of alleles per locus: 3 4 4 4 9 3 4 8
## // Number of alleles per group: 21 21 20 22 20 19 19 25 18 14 18 26
## // Percentage of missing data: 10.64 %
## // Observed heterozygosity: 0.1 0.4 0.09 0.36 0.68 0.02 0.38 0.68
## // Expected heterozygosity: 0.17 0.47 0.14 0.59 0.78 0.02 0.48 0.74</code></pre>
</div>
</div>
<div id="view-information-stored-in-genind-object" class="section level2">
<h2 class="hasAnchor">
<a href="#view-information-stored-in-genind-object" class="anchor"></a>4. View information stored in ‘genind’ object</h2>
<p>The summary lists each attribute or <code>slot</code> of the <code>genind</code> object and summarizes its content. Technically speaking, <code>genind</code> objects are S4 objects in R, which means that their slots are accessed with the <code>@</code> sign, rather than the <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> sign for attributes of the more commonly used S3 objects. Interestingly, the object does not store the data table that was imported, but converts it to a table of allele counts.</p>
<div id="a-slot-tab" class="section level3">
<h3 class="hasAnchor">
<a href="#a-slot-tab" class="anchor"></a>a) Slot <span class="citation">@tab</span>
</h3>
<p>Table with allele counts, where each column is an allele. Allele name <code>A.1</code> means “locus A, allele 1”. For a codominant marker (e.g. microsatellite) and a diploid species, allele counts per individual can be 0, 1, 2, or NA=missing. Here are the first few lines:</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="fu"><a href="https://tibble.tidyverse.org/reference/deprecated.html">as.tibble</a></span>(<span class="no">Frogs.genind</span>@<span class="kw">tab</span>)</pre></body></html></div>
<pre><code>## Warning: `as.tibble()` is deprecated as of tibble 2.0.0.
## Please use `as_tibble()` instead.
## The signature and semantics have changed, see `?as_tibble`.
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_warnings()` to see where this warning was generated.</code></pre>
<pre><code>## # A tibble: 181 x 39
##      A.1   A.2   A.3   B.1   B.3   B.2   B.4   C.1   C.2   C.4   C.5   D.1   D.2
##    &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1     2     0     0     2     0     0     0     2     0     0     0     2     0
##  2     0     2     0     2     0     0     0    NA    NA    NA    NA     2     0
##  3     2     0     0     2     0     0     0     2     0     0     0     2     0
##  4     2     0     0     2     0     0     0    NA    NA    NA    NA     0     2
##  5     1     1     0     1     1     0     0     2     0     0     0     2     0
##  6     1     1     0     2     0     0     0     2     0     0     0     1     0
##  7     0     2     0     1     1     0     0     2     0     0     0     2     0
##  8     0     2     0     1     1     0     0     2     0     0     0     2     0
##  9     1     0     1     2     0     0     0     2     0     0     0     2     0
## 10     0     2     0     1     1     0     0     2     0     0     0     2     0
## # … with 171 more rows, and 26 more variables: D.3 &lt;int&gt;, D.4 &lt;int&gt;, E.1 &lt;int&gt;,
## #   E.2 &lt;int&gt;, E.3 &lt;int&gt;, E.7 &lt;int&gt;, E.6 &lt;int&gt;, E.8 &lt;int&gt;, E.5 &lt;int&gt;,
## #   E.4 &lt;int&gt;, E.10 &lt;int&gt;, F.1 &lt;int&gt;, F.2 &lt;int&gt;, F.3 &lt;int&gt;, G.1 &lt;int&gt;,
## #   G.2 &lt;int&gt;, G.3 &lt;int&gt;, G.5 &lt;int&gt;, H.4 &lt;int&gt;, H.5 &lt;int&gt;, H.3 &lt;int&gt;,
## #   H.2 &lt;int&gt;, H.1 &lt;int&gt;, H.6 &lt;int&gt;, H.8 &lt;int&gt;, H.7 &lt;int&gt;</code></pre>
</div>
<div id="b-slot-loc-n-all" class="section level3">
<h3 class="hasAnchor">
<a href="#b-slot-loc-n-all" class="anchor"></a>b) Slot <span class="citation">@loc.n.all</span>
</h3>
<p>Number of alleles per locus, across all populations.</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="no">Frogs.genind</span>@<span class="kw">loc.n.all</span></pre></body></html></div>
<pre><code>## A B C D E F G H 
## 3 4 4 4 9 3 4 8</code></pre>
</div>
<div id="c-slot-loc-fac" class="section level3">
<h3 class="hasAnchor">
<a href="#c-slot-loc-fac" class="anchor"></a>c) Slot <span class="citation">@loc.fac</span>
</h3>
<p>This is a factor that indicates for each allele (column in <span class="citation">@tab</span>) which locus it belongs to. The levels of the factor correspond to the loci.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">Frogs.genind</span>@<span class="kw">loc.fac</span></pre></body></html></div>
<pre><code>##  [1] A A A B B B B C C C C D D D D E E E E E E E E E F F F G G G G H H H H H H H
## [39] H
## Levels: A B C D E F G H</code></pre>
</div>
<div id="d-slot-all-names" class="section level3">
<h3 class="hasAnchor">
<a href="#d-slot-all-names" class="anchor"></a>d) Slot <span class="citation">@all.names</span>
</h3>
<p>List of allele names, separately for each locus. Note: the alleles are treated as text (character), even if they are coded as numbers. They are sorted in order of occurrence in data set.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">Frogs.genind</span>@<span class="kw">all.names</span></pre></body></html></div>
<pre><code>## $A
## [1] "1" "2" "3"
## 
## $B
## [1] "1" "3" "2" "4"
## 
## $C
## [1] "1" "2" "4" "5"
## 
## $D
## [1] "1" "2" "3" "4"
## 
## $E
## [1] "1"  "2"  "3"  "7"  "6"  "8"  "5"  "4"  "10"
## 
## $F
## [1] "1" "2" "3"
## 
## $G
## [1] "1" "2" "3" "5"
## 
## $H
## [1] "4" "5" "3" "2" "1" "6" "8" "7"</code></pre>
</div>
<div id="e-further-slots" class="section level3">
<h3 class="hasAnchor">
<a href="#e-further-slots" class="anchor"></a>e) Further slots</h3>
<p>The following slots are automatically filled with default values unless specified by user during import into <code>genind</code> object:</p>
<ul>
<li>Slot <span class="citation">@ploidy</span>: ploidy for each individual.</li>
<li>Slot <span class="citation">@type</span>: codominant (microsatellites) or presence/absence (SNP, AFLP)</li>
</ul>
<p>The following slots may be empty unless specified during or after import into ‘genind’ object:</p>
<ul>
<li>Slot <span class="citation">@other</span>: placeholder for non-genetic data, e.g. spatial coordinates of individuals.</li>
<li>Slot <span class="citation">@pop</span>: vector indicating the population that each individual belongs to.</li>
<li>Slot <span class="citation">@strata</span>: stratification of populations, e.g. within regions or treatments.</li>
<li>Slot <span class="citation">@hierarchy</span>: optional formula defining hierarchical levels in strata.</li>
</ul>
</div>
</div>
<div id="import-data-with-gstudio-package" class="section level2">
<h2 class="hasAnchor">
<a href="#import-data-with-gstudio-package" class="anchor"></a>5. Import data with ‘gstudio’ package</h2>
<div id="a-function-read_population" class="section level3">
<h3 class="hasAnchor">
<a href="#a-function-read_population" class="anchor"></a>a) Function ‘read_population’</h3>
<p>While ‘genind’ objects are used by many functions for analyzing genetic data, the <code>gstudio</code> package provides an interface for additional analysis. We use the function <code>read_population</code> to import the data.</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r">?<span class="no">read_population</span></pre></body></html></div>
</div>
<div id="b-set-arguments" class="section level3">
<h3 class="hasAnchor">
<a href="#b-set-arguments" class="anchor"></a>b) Set arguments</h3>
<p>Some explanations:</p>
<ul>
<li>
<strong>path</strong>: This is the path to the .csv file. If it is in the current working directory, then the filename is sufficient.<br>
</li>
<li>
<strong>type</strong>: here we have ‘separated’ data, as the alleles at each locus are separated by a colon.</li>
<li>
<strong>locus.columns</strong>: this would be columns 3 - 10 in this case, because we are importing from the .csv file directly, which does not contain the FrogID variable. We can add that later.</li>
<li>
<strong>phased</strong>: default is fine here.</li>
<li>
<strong>sep</strong>: here this refers to the .csv file! The separater is a comma.</li>
<li>
<strong>header</strong>: yes we have column names.</li>
</ul>
<p>Working directory confusion: If you are working within an R project folder, then that folder is automatically set as your working directory. If you are working in an R session outside of an R project, then you need to actively set the working directory in the R Studio menu under: <code>Session &gt; Set working directory &gt; Choose Directory</code>.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="no">Frogs.gstudio</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gstudio/man/read_population.html">read_population</a></span>(<span class="kw">path</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"ralu.loci.csv"</span>,
                                                  <span class="kw">package</span> <span class="kw">=</span> <span class="st">"LandGenCourse"</span>),
                   <span class="kw">type</span><span class="kw">=</span><span class="st">"separated"</span>, <span class="kw">locus.columns</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>:<span class="fl">10</span>),
                   <span class="kw">phased</span><span class="kw">=</span><span class="fl">FALSE</span>, <span class="kw">sep</span><span class="kw">=</span><span class="st">","</span>, <span class="kw">header</span><span class="kw">=</span><span class="fl">TRUE</span>)</pre></body></html></div>
</div>
<div id="c-check-imported-data-1" class="section level3">
<h3 class="hasAnchor">
<a href="#c-check-imported-data-1" class="anchor"></a>c) Check imported data</h3>
<p>Check the column types: <code>SiteName</code> and <code>Pop</code> are now character vectors, and the loci columns A - H are now of class <code>locus</code> (it is indicated that ‘locus’ is a S3 object type).</p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span>(<span class="no">Frogs.gstudio</span>)</pre></body></html></div>
<pre><code>## 'data.frame':    181 obs. of  10 variables:
##  $ SiteName: chr  "AirplaneLake" "AirplaneLake" "AirplaneLake" "AirplaneLake" ...
##  $ Pop     : chr  "Airplane" "Airplane" "Airplane" "Airplane" ...
##  $ A       : 'locus' chr  "1:1" "2:2" "1:1" "1:1" ...
##  $ B       : 'locus' chr  "1:1" "1:1" "1:1" "1:1" ...
##  $ C       : 'locus' chr  "1:1" "" "1:1" "" ...
##  $ D       : 'locus' chr  "1:1" "1:1" "1:1" "2:2" ...
##  $ E       : 'locus' chr  "1:2" "1:1" "3:3" "1:2" ...
##  $ F       : 'locus' chr  "1:1" "" "1:1" "" ...
##  $ G       : 'locus' chr  "1:1" "2:2" "1:1" "" ...
##  $ H       : 'locus' chr  "4:5" "" "3:3" "" ...</code></pre>
</div>
<div id="d-add-frogid" class="section level3">
<h3 class="hasAnchor">
<a href="#d-add-frogid" class="anchor"></a>d) Add FrogID</h3>
<p><code>Frogs.gstudio</code> is a data frame, and we can manipulate like any other data frame. Here we will add the variable <code>FrogID</code> as the first column.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">Frogs.gstudio</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">FrogID</span><span class="kw">=</span><span class="no">Frogs</span>$<span class="no">FrogID</span>, <span class="no">Frogs.gstudio</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">Frogs.gstudio</span>)</pre></body></html></div>
<pre><code>##   FrogID     SiteName      Pop   A   B   C   D   E   F   G   H
## 1  Air.1 AirplaneLake Airplane 1:1 1:1 1:1 1:1 1:2 1:1 1:1 4:5
## 2  Air.2 AirplaneLake Airplane 2:2 1:1     1:1 1:1     2:2    
## 3  Air.3 AirplaneLake Airplane 1:1 1:1 1:1 1:1 3:3 1:1 1:1 3:3
## 4  Air.4 AirplaneLake Airplane 1:1 1:1     2:2 1:2            
## 5  Air.5 AirplaneLake Airplane 1:2 1:3 1:1 1:1 1:2 1:1 1:2 4:5
## 6  Air.6 AirplaneLake Airplane 1:2 1:1 1:1 1:3 1:1 1:1 1:2 4:5</code></pre>
</div>
<div id="e-convert-to-genind-object" class="section level3">
<h3 class="hasAnchor">
<a href="#e-convert-to-genind-object" class="anchor"></a>e) Convert to ‘genind’ object</h3>
<p>We can use the dataframe with the locus objects from gstudio to import the data into a genind object. In this case, we set <code>sep=":"</code>, because the <code>locus</code> object in <code>gstudio</code> stores the alleles at each locus separated by a colon, and <code>NA.char=""</code>, because <code>gstudio</code> stores missing values as empty cells.</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="no">Frogs.genind2</span> <span class="kw">&lt;-</span> <span class="kw pkg">adegenet</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/df2genind.html">df2genind</a></span>(<span class="kw">X</span><span class="kw">=</span><span class="no">Frogs.gstudio</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">4</span>:<span class="fl">11</span>)], <span class="kw">sep</span><span class="kw">=</span><span class="st">":"</span>, <span class="kw">ncode</span><span class="kw">=</span><span class="kw">NULL</span>,
                          <span class="kw">ind.names</span><span class="kw">=</span><span class="no">Frogs.gstudio</span>$<span class="no">FrogID</span>, <span class="kw">loc.names</span><span class="kw">=</span><span class="kw">NULL</span>,
                          <span class="kw">pop</span><span class="kw">=</span><span class="no">Frogs.gstudio</span>$<span class="no">Pop</span>, <span class="kw">NA.char</span><span class="kw">=</span><span class="st">""</span>, <span class="kw">ploidy</span><span class="kw">=</span><span class="fl">2</span>,
                          <span class="kw">type</span><span class="kw">=</span><span class="st">"codom"</span>, <span class="kw">strata</span><span class="kw">=</span><span class="kw">NULL</span>, <span class="kw">hierarchy</span><span class="kw">=</span><span class="kw">NULL</span>)
<span class="no">Frogs.genind2</span></pre></body></html></div>
<pre><code>## /// GENIND OBJECT /////////
## 
##  // 181 individuals; 8 loci; 39 alleles; size: 55.5 Kb
## 
##  // Basic content
##    @tab:  181 x 39 matrix of allele counts
##    @loc.n.all: number of alleles per locus (range: 3-9)
##    @loc.fac: locus factor for the 39 columns of @tab
##    @all.names: list of allele names for each locus
##    @ploidy: ploidy of each individual  (range: 2-2)
##    @type:  codom
##    @call: adegenet::df2genind(X = Frogs.gstudio[, c(4:11)], sep = ":", 
##     ncode = NULL, ind.names = Frogs.gstudio$FrogID, loc.names = NULL, 
##     pop = Frogs.gstudio$Pop, NA.char = "", ploidy = 2, type = "codom", 
##     strata = NULL, hierarchy = NULL)
## 
##  // Optional content
##    @pop: population of each individual (group size range: 7-23)</code></pre>
</div>
</div>
<div id="r-exercise-week-1" class="section level2">
<h2 class="hasAnchor">
<a href="#r-exercise-week-1" class="anchor"></a>6. R exercise Week 1</h2>
<p><strong>R Notebook:</strong> Create a new R Notebook for each weekly exercise. Watch the course video “Week 0: Intro to R Notebooks” as needed. At the end, “knit” it to an html file and view it in your browser.</p>
<p>Good news: if you can knit the file, the code can stand by itself (it does not depend on what dyou did in your R session before) and runs without errors. This is a good check. If there are error messages, check the ‘R Markdown’ tab for the code line number and try to fix it.</p>
<p><strong>Task:</strong> Import the data set <code>pulsatilla_genotypes</code>, which we’ll use in a later lab (Week 14), into <code>gstudio</code> and convert it to a <code>genind</code> object.</p>
<p><strong>Data:</strong> This file contains microsatellite data for adults and seeds of the herb <em>Pulsatilla vulgaris</em> sampled at seven sites. Reference: DiLeo et al. (2018), Journal of Ecology 106:2242-2255. <a href="https://besjournals.onlinelibrary.wiley.com/doi/10.1111/1365-2745.12992" class="uri">https://besjournals.onlinelibrary.wiley.com/doi/10.1111/1365-2745.12992</a></p>
<p>The following code copies the file into the ‘dowloads’ folder in your R project folder:</p>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"pulsatilla_genotypes.csv"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"LandGenCourse"</span>),
          <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span>(), <span class="st">"/downloads/pulsatilla_genotypes.csv"</span>), <span class="kw">overwrite</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
<pre><code>## [1] FALSE</code></pre>
<p>Variables:</p>
<ul>
<li>
<strong>ID</strong>: Family ID (i.e., mother and her offspring have the same ID)</li>
<li>
<strong>OffID</strong>: ‘0’ for adults, seeds from the same mom are numbered 1, 2, etc.</li>
<li>
<strong>Population</strong>: site ID</li>
<li>
<strong>Coordinates</strong>: X and Y coordinates (Projection info: EPSG Projection 31468)</li>
<li>
<strong>Loci</strong>: seven diploid microsatellites, each with two columns (1 allele per column)</li>
</ul>
<p><strong>Hints:</strong></p>
<ol style="list-style-type: lower-alpha">
<li>
<strong>Load packages</strong>: Make sure the following packages are loaded: <code>gstudio</code>, <code>here</code>,<code>tibble</code> and <code>adegenet</code>.</li>
<li>
<strong>View data file</strong>: Adapt the code from section 2.c to import the raw data set. The file has a header row. View it. How are the genetic data coded?</li>
<li>
<strong>Import data into gstudio</strong>: Adapt the code from section 5.b to import the genetic data with ‘gstudio’. The loci are in columns 6 - 19. What setting for <code>type</code> is appropriate for this data set? Check the help file for <code>read_population</code>.</li>
<li>
<strong>Check imported data</strong>: Use <code>str</code> or <code>as_tibble</code> to check the imported data. Does each variable have the correct data type? Note: there should be 7 variables of type <code>locus</code>.</li>
<li>
<strong>Check variable types</strong>: Create a bulleted list, like the one above, with the variables. For each variable, list their R data type (e.g., numeric, integer, character, logical, factor, locus). Check the cheatsheet “R markdown language” as needed.</li>
<li>
<strong>Convert to genind object</strong>: Modify the code from section 5.e to convert the data from <code>gstudio</code> to a <code>genind</code> object. Ignore the warning about duplicate labels. Print a summary of the <code>genind</code> object.</li>
</ol>
<p><strong>Question:</strong> What is the range of the number of genotyped individuals per population in this dataset?</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://sites.utm.utoronto.ca/wagnerlab/">Helene Wagner</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
