<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Week 4: Metapopulation Genetics • LandGenCourse</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Week 4: Metapopulation Genetics">
<meta property="og:description" content="LandGenCourse">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-168837293-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-168837293-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LandGenCourse</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Week0_BasicR.html">Week 0: Intro to R Programming</a>
    </li>
    <li>
      <a href="../articles/Week0_Graphics.html">Week 0: R Graphics</a>
    </li>
    <li>
      <a href="../articles/Week1_vignette.html">Week 1: Importing Genetic Data</a>
    </li>
    <li>
      <a href="../articles/Week2_vignette.html">Week 2: Spatial Data</a>
    </li>
    <li>
      <a href="../articles/Week2_bonus_vignette.html">Week 2: Bonus Material</a>
    </li>
    <li>
      <a href="../articles/Week3_vignette.html">Week 3: Genetic Diversity</a>
    </li>
    <li>
      <a href="../articles/Week4_vignette.html">Week 4: Metapopulation Genetics</a>
    </li>
    <li>
      <a href="../articles/Week5_vignette.html">Week 5: Spatial Statistics</a>
    </li>
    <li>
      <a href="../articles/Week6_vignette.html">Week 6: Quantitative Genetics</a>
    </li>
    <li>
      <a href="../articles/Week7_vignette.html">Week 7: Spatial Linear Models</a>
    </li>
    <li>
      <a href="../articles/Week8_vignette.html">Week 8: Simulation</a>
    </li>
    <li>
      <a href="../articles/Week8_bonus_vignette.html">Week 8: Bonus Material</a>
    </li>
    <li>
      <a href="../articles/Week9_vignette.html">Week 9: Population Structure</a>
    </li>
    <li>
      <a href="../articles/Week10_vignette.html">Week 10: Landscape Resistance</a>
    </li>
    <li>
      <a href="../articles/Week11_vignette.html">Week 11: Detecting Adaptation</a>
    </li>
    <li>
      <a href="../articles/Week12_vignette.html">Week 12: Model Selection</a>
    </li>
    <li>
      <a href="../articles/Week13_vignette.html">Week 13: Gravity Models</a>
    </li>
    <li>
      <a href="../articles/Week14_vignette.html">Week 14: Contemporary Gene Flow</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../tutorials/00-Intro_RStudio.html">Introduction to RStudio</a>
    </li>
    <li>
      <a href="../tutorials/00-Intro_LandGenCourse.html">Introduction to the LandGenCourse package</a>
    </li>
    <li>
      <a href="../tutorials/00-Intro_RNotebooks.html">Introduction to R Notebooks</a>
    </li>
    <li>
      <a href="../tutorials/00-Version_Control.html">Version Control 101</a>
    </li>
    <li>
      <a href="../tutorials/01-Genetic_Data.html">Week 1 Getting Landscape Genetic Data into R</a>
    </li>
    <li>
      <a href="../tutorials/02-Spatial_Data.html">Week 2 Land Cover Analysis in R</a>
    </li>
    <li>
      <a href="../tutorials/03-Genetic_Diversity.html">Week 3 Genetic Diversity (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/03-Genetic_Diversity_Part2.html">Week 3 Genetic Diversity (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/04-Metapop_Genetics.html">Week 4 Metapopulation Genetics (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/04-Metapop_Genetics_Part2.html">Week 4 Metapopulation Genetics (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/05-Spatial_Stats.html">Week 5 Spatial Statistics</a>
    </li>
    <li>
      <a href="../tutorials/06-Quantitative_Genetics.html">Week 6 Quantitative Genetics (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/06-Quantitative_Genetics_Part2.html">Week 6 Quantitative Genetics (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/07-Spatial_Models.html">Week 7 Spatial Linear Models (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/07-Spatial_Models_Part2.html">Week 7 Spatial Linear Models (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/08-Modeling.html">Week 8 Simulation Experiments in R (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/08-Modeling_Part2.html">Week 8 Simulation Experiments in R (Part 2)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Week 4: Metapopulation Genetics</h1>
                        <h4 class="author">Nusha Keyghobadi, Thomas Lamy and Helene Wagner</h4>
            
            <h4 class="date">2020-06-08</h4>
      
      
      <div class="hidden name"><code>Week4_vignette.Rmd</code></div>

    </div>

    
    
<div id="overview-of-worked-example" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-of-worked-example" class="anchor"></a>1. Overview of Worked Example</h2>
<div id="a-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#a-goals" class="anchor"></a>a) Goals</h3>
<p>This worked example shows how to:</p>
<ul>
<li>Assess the spatial distribution of genetic structure in a metapopulation using hierarchical AMOVA</li>
<li>Relate site-specific Fst to patch connectivity and population size</li>
<li>Relate site-specific genetic diversity to explanatory variables (node-level analysis)</li>
<li>Assess temporal changes (between years for same site) and evidence for extinction events</li>
<li>Perform power analysis and sample size calculation for the temporal study</li>
</ul>
</div>
<div id="b-data-set" class="section level3">
<h3 class="hasAnchor">
<a href="#b-data-set" class="anchor"></a>b) Data set</h3>
<p>Lamy et al. (2012) sampled the freshwater snail <em>Drepanotrema depressissimum</em> in a fragmented landscape of tropical ponds on the island of Guadeloupe in the French West Indies. They used a spatially and temporally stratified sampling design with a total of 25 sites, where 12 sites formed four well-separated clusters of three neighbouring sites each, to study spatial variability, and 12 sites spread across the island were sampled in multiple years to study temporal variability. For each site and year, 22 - 34 individuals were genotyped at ten microsatellite loci. The species is diploid, hermaphroditic, and outcrossed.</p>
<p>A key characteristic of this system is the presence of a dry and a rainy season. In the dry season, many ponds can dry out, possibly causing extinction of the local snail populations. During the rainy season, ponds refill and can even overflow, thereby becoming connected through the hydrological network. During this rainy season, dispersal between ponds may occur.</p>
<ul>
<li><p><strong>dd.ecogen</strong>: The dataset ‘dd.ecogen’ with genetic data for 1270 snails from 42 populations is included in package ‘LandGenCourse’. To load it, type: data(dd.ecogen).</p></li>
<li><p><strong>dd.site</strong>: Population-level data from Tables 2 - 5 of Lamy et al. (2012) are available in dataset ‘dd.site’ (with 25 variables) in package ‘LandGenCourse’. To load it, type: data(dd.site).</p></li>
</ul>
<div id="reference" class="section level4">
<h4 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference:</h4>
<p>Lamy, T., Pointier, J. P., Jarne, P. and David, P. (2012), Testing metapopulation dynamics using genetic, demographic and ecological data. Molecular Ecology, 21: 1394–1410. <a href="doi:10.1111/j.1365-294X.2012.05478.x" class="uri">doi:10.1111/j.1365-294X.2012.05478.x</a></p>
</div>
</div>
<div id="c-required-r-libraries" class="section level3">
<h3 class="hasAnchor">
<a href="#c-required-r-libraries" class="anchor"></a>c) Required R libraries</h3>
<p>All required packages should have been installed already when you installed ‘LandGenCourse’.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">LandGenCourse</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">EcoGenetics</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">methods</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="co">#require(tibble)</span>
<span class="co">#require(poppr)</span>
<span class="co">#require(ade4)</span>
<span class="co">#require(pwr)</span>
<span class="co">#require(effsize)</span>
<span class="co">#require(sp)</span>
<span class="co">#require(ggmap)</span>
<span class="co">#require(car)  </span></pre></body></html></div>
</div>
<div id="d-import-data" class="section level3">
<h3 class="hasAnchor">
<a href="#d-import-data" class="anchor"></a>d) Import data</h3>
<p>Let’s import the genetic data (spatial and temporal data sets combined, 42 combinations of site and year). Use ‘?dd.ecogen’ to check helpfile with data set desription.</p>
<p>The ‘ecogen’ object ‘dd.ecogen’ contains individual-level data in the following slots:</p>
<ul>
<li>
<strong>XY</strong>: Spatial coordinates (lat-long format)</li>
<li>
<strong>G</strong>: Microsatellite loci (columns = loci, rows = individuals)</li>
<li>
<strong>A</strong>: Table of allele frequencies (columns = alleles, rows = individuals)</li>
<li>
<strong>S</strong>: Structure variables (SiteID, SITE, YEAR, Cluster)</li>
</ul>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">dd.ecogen</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"LandGenCourse"</span>)
<span class="no">dd.ecogen</span></pre></body></html></div>
<pre><code>## 
##                    || ECOGEN CLASS OBJECT ||
## ----------------------------------------------------------------------------
##  Access to slots: &lt;ecoslot.&gt; + &lt;name of the slot&gt; + &lt;(name of the object)&gt; 
##  See: help("EcoGenetics accessors")
## ----------------------------------------------------------------------------
## 
##  | slot XY:    | --&gt; 1270 x 2       coordinates
##  | slot P:     | --&gt; 0 x 0          phenotypic variables
##  | slot G:     | --&gt; 1270 x 10      loci         &gt;&gt; ploidy: 2 || codominant
##  | slot A:     | --&gt; 1270 x 372     alleles
##  | slot E:     | --&gt; 0 x 0          environmental variables
##  | slot S:     | --&gt; 1270 x 4       structures   &gt;&gt; 4 structures found
##  | slot C:     | --&gt; 0 x 0          variables
##  | slot OUT:   | --&gt; 0              results
## ----------------------------------------------------------------------------</code></pre>
<div class="sourceCode" id="cb4"><html><body><pre class="r">?<span class="no">dd.ecogen</span></pre></body></html></div>
<p>We also import site-level data from Tables 2 - 5 in Lamy et al. (2012). Use ‘?dd.site’ to check helpfile with data set desription of the variables.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">dd.site</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"LandGenCourse"</span>)
<span class="kw pkg">tibble</span><span class="kw ns">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span>(<span class="no">dd.site</span>)</pre></body></html></div>
<pre><code>## # A tibble: 42 x 24
##    SiteID SITE   YEAR Spatial MultiYear APE   Cluster     n    RA    He      f
##    &lt;chr&gt;  &lt;fct&gt; &lt;int&gt; &lt;lgl&gt;   &lt;lgl&gt;     &lt;lgl&gt; &lt;fct&gt;   &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
##  1 PICO2… PICO   2006 FALSE   TRUE      FALSE &lt;NA&gt;       34  11.2 0.897  0.038
##  2 PICO2… PICO   2007 FALSE   TRUE      FALSE &lt;NA&gt;       24  11.4 0.911  0.022
##  3 PICO2… PICO   2009 TRUE    TRUE      FALSE &lt;NA&gt;       32  11.4 0.896  0.024
##  4 ROC20… ROC    2006 TRUE    TRUE      FALSE &lt;NA&gt;       32  11.2 0.831  0.005
##  5 ROC20… ROC    2007 FALSE   TRUE      FALSE &lt;NA&gt;       32  11.0 0.82   0.018
##  6 SEN20… SEN    2006 FALSE   TRUE      FALSE &lt;NA&gt;       32  11.2 0.876 -0.008
##  7 SEN20… SEN    2007 TRUE    TRUE      FALSE &lt;NA&gt;       30  11.6 0.88   0.046
##  8 VEE20… VEE    2006 FALSE   TRUE      FALSE &lt;NA&gt;       31  10.5 0.864  0.034
##  9 VEE20… VEE    2007 FALSE   TRUE      FALSE &lt;NA&gt;       32  11.3 0.854  0.044
## 10 VEE20… VEE    2008 TRUE    TRUE      FALSE &lt;NA&gt;       31  11.2 0.871  0.014
## # … with 32 more rows, and 13 more variables: s &lt;dbl&gt;, Type &lt;fct&gt;,
## #   FST.GESTE &lt;dbl&gt;, Size &lt;dbl&gt;, V &lt;dbl&gt;, C &lt;dbl&gt;, Stab &lt;dbl&gt;, D &lt;int&gt;,
## #   APA &lt;int&gt;, NLT &lt;dbl&gt;, Fst.temp &lt;dbl&gt;, Longitude &lt;dbl&gt;, Latitude &lt;dbl&gt;</code></pre>
<div class="sourceCode" id="cb7"><html><body><pre class="r">?<span class="no">dd.site</span></pre></body></html></div>
<p><strong>Questions</strong>: with the help file for ‘dd.site’, check the meaning of the following explanatory variables:</p>
<ul>
<li>What does ‘APE’ refer to, and how is it different from ‘APA’?</li>
<li>What does ‘NLT’ represent, and is it calculated independently from ‘Size’?</li>
<li>What does ‘Type’ mean, and what about ‘V’ and ‘D’?</li>
<li>To understand how connectivity ‘C’ and stability ‘Stab’ were calculated, you’ll need to consult Lamy et al. (2012).</li>
</ul>
<p><strong>Your hypothesis</strong>: which explanatory variables would you expect to affect:</p>
<ul>
<li>Genetic diversity within local populations?</li>
<li>Genetic differentiation among local populations?</li>
<li>Both?</li>
</ul>
<p>In the following, we’ll perform three types of analyses:</p>
<ul>
<li>Compare 25 populations in space, across the island of Guadeloupe.</li>
<li>Compare 12 populations in 4 clusters: differentiation within vs. among clusters?</li>
<li>Compare 12 sites over time, some of which experienced a local extinction event.</li>
</ul>
</div>
</div>
<div id="spatial-distribution-of-genetic-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#spatial-distribution-of-genetic-structure" class="anchor"></a>2. Spatial distribution of genetic structure</h2>
<p>How similar are populations from nearby habitat patches compared to populations across the island? To answer this question, we perform a hiearchical AMOVA (analysis of molecular variance) with individuals from 12 populations that form 4 clusters with 3 populations each.</p>
<div id="a-creating-a-genind-object-with-the-hierarchical-data-set" class="section level3">
<h3 class="hasAnchor">
<a href="#a-creating-a-genind-object-with-the-hierarchical-data-set" class="anchor"></a>a) Creating a ‘genind’ object with the hierarchical data set</h3>
<p>First, we need to extract the samples that belong to the hierarchical data set. There are four clusters: “North”, “East”, “Center” and “South”. We are looking for the observations where the variable “Cluster” has one of these four values, all other observations will have a missing value for “Cluster”. We can use ‘!is.na’ to extract all rows with non-missing values. Then we convert to a ‘genind’ object.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">dd.ecogen.Cluster</span> <span class="kw">&lt;-</span> <span class="no">dd.ecogen</span>[!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">dd.ecogen</span><span class="kw">[[</span><span class="st">"S"</span>]]$<span class="no">Cluster</span>),]
<span class="no">dd.genind.Cluster</span> <span class="kw">&lt;-</span> <span class="kw pkg">EcoGenetics</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/EcoGenetics/man/ecogen2genind.html">ecogen2genind</a></span>(<span class="no">dd.ecogen.Cluster</span>)</pre></body></html></div>
<pre><code>## Loading required package: adegenet</code></pre>
<pre><code>## Loading required package: ade4</code></pre>
<pre><code>## Registered S3 method overwritten by 'spdep':
##   method   from
##   plot.mst ape</code></pre>
<pre><code>## 
##    /// adegenet 2.1.3 is loaded ////////////
## 
##    &gt; overview: '?adegenet'
##    &gt; tutorials/doc/questions: 'adegenetWeb()' 
##    &gt; bug reports/feature requests: adegenetIssues()</code></pre>
</div>
<div id="b-hierarchical-amova" class="section level3">
<h3 class="hasAnchor">
<a href="#b-hierarchical-amova" class="anchor"></a>b) Hierarchical AMOVA</h3>
<p>There are several implementations of AMOVA in R, e.g. in pacakges ‘ade4’, ‘pegas’ and ‘vegan’. The ‘ade4’ implementation is closest to the original implementation in Arlequin. Package ‘poppr’ has a wrapper function ‘poppr.amova’ that makes it easy to perform AMOVA with the ‘ade4’ or with the ‘pegas’ implementation (see ‘?poppr.amova’ for a discussion of their pros and cons). Here we’ll use ‘ade4’.</p>
<ul>
<li>The first argument is the ‘genind’ object.</li>
<li>The argument ‘hier’ defines the hierarchy, with the top level first (i.e., here SITE is nested within Cluster). The variables are expected to be found in the <span class="citation">@strata</span> slot of the ‘genind’ object.</li>
<li>The argument ‘within=FALSE’ specifies that within-individual variance (i.e., observed heterozygosity) should not be tested. Setting this to ‘TRUE’ can lead to problems with missing values.</li>
</ul>
<p>First we run the AMOVA and estimate the percent of molecular variance at each hierarchical level.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">amova.result</span> <span class="kw">&lt;-</span> <span class="kw pkg">poppr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/poppr/man/poppr.amova.html">poppr.amova</a></span>(<span class="no">dd.genind.Cluster</span>, <span class="kw">hier</span> <span class="kw">=</span> ~ <span class="no">Cluster</span>/<span class="no">SITE</span>,
                                   <span class="kw">within</span><span class="kw">=</span><span class="fl">FALSE</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"ade4"</span>)</pre></body></html></div>
<pre><code>## Registered S3 method overwritten by 'pegas':
##   method      from
##   print.amova ade4</code></pre>
<pre><code>## 
##  No missing values detected.</code></pre>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">amova.result</span></pre></body></html></div>
<pre><code>## $call
## ade4::amova(samples = xtab, distances = xdist, structures = xstruct)
## 
## $results
##                                 Df     Sum Sq   Mean Sq
## Between Cluster                  3   67.27429 22.424764
## Between samples Within Cluster   8  158.37917 19.797396
## Within samples                 353 1563.08216  4.427995
## Total                          364 1788.73562  4.914109
## 
## $componentsofcovariance
##                                                 Sigma           %
## Variations  Between Cluster                0.02770227   0.5582719
## Variations  Between samples Within Cluster 0.50644882  10.2062461
## Variations  Within samples                 4.42799478  89.2354820
## Total variations                           4.96214587 100.0000000
## 
## $statphi
##                             Phi
## Phi-samples-total   0.107645180
## Phi-samples-Cluster 0.102635446
## Phi-Cluster-total   0.005582719</code></pre>
<p>Then we test whether each variance component is statistically significant (i.e., significantly larger than zero).</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">amova.test</span> <span class="kw">&lt;-</span> <span class="kw pkg">ade4</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ade4/man/randtest.html">randtest</a></span>(<span class="no">amova.result</span>, <span class="kw">nrepet</span> <span class="kw">=</span> <span class="fl">999</span>)
<span class="no">amova.test</span></pre></body></html></div>
<pre><code>## class: krandtest lightkrandtest 
## Monte-Carlo tests
## Call: randtest.amova(xtest = amova.result, nrepet = 999)
## 
## Number of tests:   3 
## 
## Adjustment method for multiple comparisons:   none 
## Permutation number:   999 
##                         Test        Obs   Std.Obs   Alter Pvalue
## 1  Variations within samples 4.42799478 -84.81343    less  0.001
## 2 Variations between samples 0.50644882  55.27578 greater  0.001
## 3 Variations between Cluster 0.02770227   1.65383 greater  0.062</code></pre>
<p><strong>Questions</strong>:</p>
<ul>
<li>At what level is there stronger differentiation, within or among clusters?</li>
<li>What does this mean biologically?</li>
<li>Are both levels statistically significant?</li>
</ul>
</div>
</div>
<div id="what-determines-genetic-differentiation-among-sites" class="section level2">
<h2 class="hasAnchor">
<a href="#what-determines-genetic-differentiation-among-sites" class="anchor"></a>3. What determines genetic differentiation among sites?</h2>
<p>What factors explain site-specific Fst? Let’s consider the key micro-evolutionary processes:</p>
<ul>
<li>
<strong>Genetic drift</strong>: the smaller the population, the higher the rate of drift, hence we expect higher differention for small populations. Predictor: long-term population size ‘NLT’.</li>
<li>
<strong>Gene flow</strong>: gene flow homogenizes allele frequencies, hence we expect less differentiation for well connected patches. Predictors: connectivity ‘C’, density of favorable habitat ‘D’ (within 2 km radius).</li>
</ul>
<p>First, we create a new SpatialPointsDataFrame with the subset of data for the spatial analysis (25 ponds, one year each).</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">dd.spatial</span> <span class="kw">&lt;-</span> <span class="no">dd.site</span>[<span class="no">dd.site</span>@<span class="kw">data</span>$<span class="no">Spatial</span><span class="kw">==</span><span class="fl">TRUE</span>, ]</pre></body></html></div>
<div id="a-correlation-matrix" class="section level3">
<h3 class="hasAnchor">
<a href="#a-correlation-matrix" class="anchor"></a>a) Correlation matrix</h3>
<p>Let’s start with a correlation matrix.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="no">dd.spatial</span>@<span class="kw">data</span>[ , <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"FST.GESTE"</span>, <span class="st">"NLT"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>)],
    <span class="kw">use</span><span class="kw">=</span><span class="st">"pairwise.complete"</span>)</pre></body></html></div>
<pre><code>##             FST.GESTE         NLT           C           D
## FST.GESTE  1.00000000 -0.48565247 -0.43104600  0.02768978
## NLT       -0.48565247  1.00000000 -0.08369541  0.10242973
## C         -0.43104600 -0.08369541  1.00000000 -0.32078070
## D          0.02768978  0.10242973 -0.32078070  1.00000000</code></pre>
<p><strong>Questions</strong>:</p>
<ul>
<li>Is there genetic evidence for higher drift in small populations?</li>
<li>Is there genetic evidence for higher gene flow among well connected patches?</li>
<li>Are the two factors confounded for this data set?</li>
<li>Would you prefer ‘C’ or ‘D’ to quantify patch connectivity? Does it matter? What does this mean biologically?</li>
</ul>
</div>
<div id="b-scatterplots" class="section level3">
<h3 class="hasAnchor">
<a href="#b-scatterplots" class="anchor"></a>b) Scatterplots</h3>
<p>Let’s plot the response variable FST.GESTE against each of the two predictors NLT and C. Here, we use functions from the package <code>ggplot2</code> (already loaded) to define two ggplot objects <code>NLT.plot</code> and <code>C.plot</code>, then we plot them side-by-side with the function <code><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">cowplot::plot_grid</a></code>.</p>
<ul>
<li>For NLT.plot, we define the dataset as <code>dd.spatial@data</code>, the x-axis as variable NLT, the y-axis as variable FST.GESTE, and the labels as variable SITE.</li>
<li>We add points with geom_point.</li>
<li>We add a regression line (<code>geom_smooth</code>), make it linear (<code>method = lm</code>) and add a shaded area for plus/minus 1 SE of the mean for a given value of x (<code>se = TRUE</code>).</li>
<li>We add backfilled labels (<code>geom_label</code>), define their size (<code>size</code>), and move them up a little along the y-axis (<code>nudge_y</code>)</li>
</ul>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="no">NLT.plot</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">dd.spatial</span>@<span class="kw">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">NLT</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">FST.GESTE</span>, <span class="kw">label</span><span class="kw">=</span><span class="no">SITE</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span>(<span class="kw">method</span> <span class="kw">=</span> <span class="no">lm</span>, <span class="kw">se</span> <span class="kw">=</span> <span class="fl">TRUE</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span>(<span class="kw">size</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">nudge_x</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">nudge_y</span><span class="kw">=</span><span class="fl">0.01</span>, <span class="kw">check_overlap</span><span class="kw">=</span><span class="fl">TRUE</span>)

<span class="no">C.plot</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">dd.spatial</span>@<span class="kw">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">C</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">FST.GESTE</span>, <span class="kw">label</span><span class="kw">=</span><span class="no">SITE</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span>(<span class="kw">method</span> <span class="kw">=</span> <span class="no">lm</span>, <span class="kw">se</span> <span class="kw">=</span> <span class="fl">TRUE</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span>(<span class="kw">size</span><span class="kw">=</span><span class="fl">2.5</span>, <span class="kw">nudge_x</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">nudge_y</span><span class="kw">=</span><span class="fl">0.01</span>, <span class="kw">check_overlap</span><span class="kw">=</span><span class="fl">TRUE</span>)

<span class="kw pkg">cowplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(<span class="no">NLT.plot</span>, <span class="no">C.plot</span>)</pre></body></html></div>
<pre><code>## `geom_smooth()` using formula 'y ~ x'
## `geom_smooth()` using formula 'y ~ x'</code></pre>
<p><img src="Week4_vignette_files/figure-html/unnamed-chunk-9-1.png" width="768"></p>
</div>
<div id="c-regression-model" class="section level3">
<h3 class="hasAnchor">
<a href="#c-regression-model" class="anchor"></a>c) Regression model</h3>
<p>The two predictors ‘NLT’ and ‘C’ are not strongly correlated. We’ll fit a regression model with both predictors. Here we use function ‘scale’ to standardize each variable, so that we can interpret the regression slope coefficients as partial correlation coefficients (beta coefficients).</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">mod.diff</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">FST.GESTE</span>) ~ <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">NLT</span>) + <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">C</span>),
               <span class="kw">data</span><span class="kw">=</span><span class="no">dd.spatial</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">mod.diff</span>)</pre></body></html></div>
<pre><code>## 
## Call:
## lm(formula = scale(FST.GESTE) ~ scale(NLT) + scale(C), data = dd.spatial)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -1.0576 -0.4142 -0.1361  0.2653  2.2394 
## 
## Coefficients:
##               Estimate Std. Error t value Pr(&gt;|t|)   
## (Intercept) -2.576e-17  1.535e-01   0.000  1.00000   
## scale(NLT)  -5.254e-01  1.572e-01  -3.342  0.00296 **
## scale(C)    -4.750e-01  1.572e-01  -3.021  0.00628 **
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 0.7676 on 22 degrees of freedom
## Multiple R-squared:  0.4599, Adjusted R-squared:  0.4108 
## F-statistic: 9.367 on 2 and 22 DF,  p-value: 0.00114</code></pre>
<p>Is the model valid? Let’s check the residual plots:</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,<span class="fl">2</span>))
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">mod.diff</span>, <span class="kw">labels.id</span> <span class="kw">=</span> <span class="no">dd.spatial</span>@<span class="kw">data</span>$<span class="no">SITE</span>)</pre></body></html></div>
<p><img src="Week4_vignette_files/figure-html/unnamed-chunk-11-1.png" width="768"></p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">1</span>))</pre></body></html></div>
<p>If we had more than two predictors, it would be a good idea to calculate variance inflation factors. The package ‘car’ has a function ‘vif’ that takes as argument a fitted model. Here, both predictors have VIF = 1.007, which indicates no collinearity.</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="kw pkg">car</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/car/man/vif.html">vif</a></span>(<span class="no">mod.diff</span>)</pre></body></html></div>
<pre><code>## scale(NLT)   scale(C) 
##   1.007054   1.007054</code></pre>
</div>
<div id="d-which-populations-dont-fit-the-general-pattern" class="section level3">
<h3 class="hasAnchor">
<a href="#d-which-populations-dont-fit-the-general-pattern" class="anchor"></a>d) Which populations don’t fit the general pattern?</h3>
<p>Let’s plot the residuals in space. The function ‘bubble’ from the package ‘sp’ evaluates the projection information of the SpatialPointsDataFrame ‘dd.spatial’.</p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="no">dd.spatial</span>@<span class="kw">data</span>$<span class="no">Residuals</span> <span class="kw">&lt;-</span> <span class="no">mod.diff</span>$<span class="no">residuals</span>
<span class="kw pkg">sp</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/sp/man/bubble.html">bubble</a></span>(<span class="no">dd.spatial</span>, <span class="kw">zcol</span> <span class="kw">=</span> <span class="st">"Residuals"</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"red"</span>, <span class="st">"blue"</span>))</pre></body></html></div>
<p><img src="Week4_vignette_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>Or on a map from the internet, using ‘qmplot’ from the ‘ggmap’ package. It expects lat-lon coordinates that are stored in a data frame, not a SpatialPointsDataFrame. If we convert ‘dd.spatial’ with the function ‘as.data.frame’, R will return a data frame with the site variables and with the coordinates appended as additional columns. The code below does the following (see Week 4 video, part 2):</p>
<ul>
<li>Creates an index <code>a</code> that lists identifies the two potential outliers by their SITE names (logical),</li>
<li>Creates a vectors <code>a2</code> that contains the row numbers of the potential outliers,</li>
<li>Grabs a grayscale map from the internet, plots all ponds with size and color according to their residuals, and stores the resulting map in the object <code>myMap</code>.</li>
<li>Plots the map and adds labels for the two potential outliers.The argument <code>size</code> affects font size and <code>label.padding</code> determines the size of the textbox for the label.</li>
</ul>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="no">a</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html">is.element</a></span>(<span class="no">dd.spatial</span>@<span class="kw">data</span>$<span class="no">SITE</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"DESB"</span>, <span class="st">"PTC"</span>))
<span class="no">a2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">dd.spatial</span>@<span class="kw">data</span>))[<span class="no">a</span>]
<span class="no">myMap</span> <span class="kw">&lt;-</span> <span class="kw pkg">ggmap</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggmap/man/qmplot.html">qmplot</a></span>(<span class="no">Longitude</span>, <span class="no">Latitude</span>,  <span class="kw">data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">dd.spatial</span>),
              <span class="kw">source</span> <span class="kw">=</span> <span class="st">"stamen"</span>, <span class="kw">maptype</span> <span class="kw">=</span> <span class="st">"toner-lite"</span>, <span class="kw">force</span><span class="kw">=</span><span class="fl">TRUE</span>,
              <span class="kw">col</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span>(<span class="no">Residuals</span>), <span class="kw">size</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">Residuals</span>))
<span class="no">myMap</span> + <span class="kw pkg">ggplot2</span><span class="kw ns">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_label</a></span>(<span class="kw">data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">dd.spatial</span>[<span class="no">a2</span>,]),
                   <span class="kw">mapping</span> <span class="kw">=</span> <span class="kw pkg">ggplot2</span><span class="kw ns">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">Longitude</span>, <span class="no">Latitude</span>, <span class="kw">label</span> <span class="kw">=</span> <span class="no">SITE</span>),
                   <span class="kw">size</span> <span class="kw">=</span> <span class="fl">4</span>, <span class="kw">label.padding</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/reexports.html">unit</a></span>(<span class="fl">0.2</span>, <span class="st">"lines"</span>),
                   <span class="kw">col</span> <span class="kw">=</span> <span class="st">"black"</span>, <span class="kw">vjust</span> <span class="kw">=</span> <span class="fl">0</span>, <span class="kw">nudge_x</span> <span class="kw">=</span> <span class="fl">0</span>, <span class="kw">nudge_y</span> <span class="kw">=</span> -<span class="fl">0.015</span>)</pre></body></html></div>
<p><img src="Week4_vignette_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<p>Exports the last plot as a ‘png’ (Portable Network Graphics) file, specifying canvas size and resolution. To run the code, uncomment by removing ‘#’ at the beginning of each line.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r">#require(here)
#if(!dir.exists(paste0(here(),"/output"))) dir.create(paste0(here(),"/output"))
#ggplot2::ggsave(paste0(here(),"/output/ResidualMap.png"), 
#               width = 7, height = 5.5, units = "in",  dpi = 300)</pre></body></html></div>
<p>What might explain the large residuals for the two sites ‘PTC’ and ‘DESB’?</p>
<ul>
<li>Site ‘PTC’ lies on the tip of a peninsula and thus is very isolated geographically.</li>
<li>Site ‘DESB’ is a very instable site that can frequently dry out during the dry season, as it is shallow and lies in the comparatively dry northern part of the island. In addition, although ‘DESB’ is surrounded by many ponds, these ponds never get connected to ‘DESB’ hydrologically during the rainy season. Therefore, immigration can only occur via cattle or birds, which are much less important drivers of gene flow than immigration by hydrological connectivity during the rainy season.</li>
</ul>
</div>
<div id="e-regression-model-without-outliers" class="section level3">
<h3 class="hasAnchor">
<a href="#e-regression-model-without-outliers" class="anchor"></a>e) Regression model without outliers</h3>
<p>We can use the same index ‘a’ to exclude the potential outliers from the regression model:</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">mod.diff.minus2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">FST.GESTE</span>) ~ <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">NLT</span>) + <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">C</span>),
               <span class="kw">data</span><span class="kw">=</span><span class="no">dd.spatial</span>[-<span class="no">a2</span>,])
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">mod.diff.minus2</span>)</pre></body></html></div>
<pre><code>## 
## Call:
## lm(formula = scale(FST.GESTE) ~ scale(NLT) + scale(C), data = dd.spatial[-a2, 
##     ])
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -1.2815 -0.5002 -0.1009  0.4732  1.7142 
## 
## Coefficients:
##               Estimate Std. Error t value Pr(&gt;|t|)   
## (Intercept)  3.490e-17  1.562e-01   0.000  1.00000   
## scale(NLT)  -5.096e-01  1.622e-01  -3.142  0.00514 **
## scale(C)    -5.761e-01  1.622e-01  -3.552  0.00200 **
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 0.7493 on 20 degrees of freedom
## Multiple R-squared:  0.4896, Adjusted R-squared:  0.4386 
## F-statistic: 9.593 on 2 and 20 DF,  p-value: 0.001199</code></pre>
<ul>
<li>Did omitting the two sites improve model fit?</li>
<li>Did it change the nature of the results?</li>
<li>Does this affect the biologial interpretation?</li>
</ul>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,<span class="fl">2</span>))
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">mod.diff.minus2</span>, <span class="kw">labels.id</span> <span class="kw">=</span> <span class="no">dd.spatial</span>@<span class="kw">data</span>$<span class="no">SITE</span>[-<span class="no">a2</span>])</pre></body></html></div>
<p><img src="Week4_vignette_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">1</span>))</pre></body></html></div>
</div>
</div>
<div id="what-determines-genetic-diversity" class="section level2">
<h2 class="hasAnchor">
<a href="#what-determines-genetic-diversity" class="anchor"></a>4. What determines genetic diversity?</h2>
<ul>
<li>Can the same predictors (population size and connectivity) explain genetic diversity?</li>
<li>Is patch size (‘Size’) a good proxy for population size (as often used in ecological studies)?</li>
<li>Which measure of genetic diversity shows the stronger response, allelic richness (rarefied) or expected heterozygosity?</li>
</ul>
<div id="a-correlation-matrix-1" class="section level3">
<h3 class="hasAnchor">
<a href="#a-correlation-matrix-1" class="anchor"></a>a) Correlation matrix</h3>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="no">dd.spatial</span>@<span class="kw">data</span>[, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"RA"</span>, <span class="st">"He"</span>, <span class="st">"Size"</span>, <span class="st">"NLT"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>)],
    <span class="kw">use</span><span class="kw">=</span><span class="st">"pairwise.complete"</span>)</pre></body></html></div>
<pre><code>##               RA          He       Size         NLT           C           D
## RA    1.00000000  0.95151016  0.3482453  0.51343861  0.46576979 -0.09856668
## He    0.95151016  1.00000000  0.3057086  0.49851503  0.45206733 -0.09593604
## Size  0.34824532  0.30570855  1.0000000  0.76128271 -0.17765815  0.30002700
## NLT   0.51343861  0.49851503  0.7612827  1.00000000 -0.08369541  0.10242973
## C     0.46576979  0.45206733 -0.1776581 -0.08369541  1.00000000 -0.32078070
## D    -0.09856668 -0.09593604  0.3000270  0.10242973 -0.32078070  1.00000000</code></pre>
<p><strong>Questions</strong>:</p>
<ul>
<li>How strongly are the two diversity measures ‘RA’ and ‘He’ correlated?</li>
<li>Are ‘NLT’ and ‘C’ strongly correlated with the diversity measures ‘RA’ and ‘He’?</li>
<li>Is the correlation with ‘Size’ similarly strong as the correlation with ‘NLT’?</li>
<li>How strongly are ‘Size’ and ‘NLT’ correlated with each other?</li>
<li>Does ‘D’ show a stronger correlation with diversity than with differentiation?</li>
</ul>
</div>
<div id="b-regression-models" class="section level3">
<h3 class="hasAnchor">
<a href="#b-regression-models" class="anchor"></a>b) Regression models</h3>
<p>For allelic richness:</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="no">mod.RA</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">RA</span>) ~ <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">NLT</span>) + <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">C</span>), <span class="kw">data</span> <span class="kw">=</span> <span class="no">dd.spatial</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">mod.RA</span>)</pre></body></html></div>
<pre><code>## 
## Call:
## lm(formula = scale(RA) ~ scale(NLT) + scale(C), data = dd.spatial)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -1.5115 -0.5140  0.2356  0.4995  1.2462 
## 
## Coefficients:
##               Estimate Std. Error t value Pr(&gt;|t|)   
## (Intercept) -1.238e-16  1.441e-01   0.000  1.00000   
## scale(NLT)   5.563e-01  1.476e-01   3.770  0.00106 **
## scale(C)     5.123e-01  1.476e-01   3.472  0.00217 **
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 0.7204 on 22 degrees of freedom
## Multiple R-squared:  0.5243, Adjusted R-squared:  0.481 
## F-statistic: 12.12 on 2 and 22 DF,  p-value: 0.0002825</code></pre>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,<span class="fl">2</span>))
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">mod.RA</span>, <span class="kw">labels.id</span> <span class="kw">=</span> <span class="no">dd.spatial</span>@<span class="kw">data</span>$<span class="no">SITE</span>)</pre></body></html></div>
<p><img src="Week4_vignette_files/figure-html/unnamed-chunk-20-1.png" width="768"></p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">1</span>))</pre></body></html></div>
<p>For gene diversity (expected heterozygosity):</p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="no">mod.He</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">He</span>) ~ <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">NLT</span>) + <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">C</span>), <span class="kw">data</span> <span class="kw">=</span> <span class="no">dd.spatial</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">mod.He</span>)</pre></body></html></div>
<pre><code>## 
## Call:
## lm(formula = scale(He) ~ scale(NLT) + scale(C), data = dd.spatial)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -1.7545 -0.3619  0.1211  0.4458  1.2550 
## 
## Coefficients:
##               Estimate Std. Error t value Pr(&gt;|t|)   
## (Intercept) -1.381e-16  1.486e-01   0.000  1.00000   
## scale(NLT)   5.401e-01  1.522e-01   3.549  0.00180 **
## scale(C)     4.973e-01  1.522e-01   3.268  0.00352 **
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 0.7429 on 22 degrees of freedom
## Multiple R-squared:  0.4941, Adjusted R-squared:  0.4481 
## F-statistic: 10.74 on 2 and 22 DF,  p-value: 0.0005559</code></pre>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,<span class="fl">2</span>))
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">mod.He</span>, <span class="kw">labels.id</span> <span class="kw">=</span> <span class="no">dd.spatial</span>@<span class="kw">data</span>$<span class="no">SITE</span>)</pre></body></html></div>
<p><img src="Week4_vignette_files/figure-html/unnamed-chunk-22-1.png" width="768"></p>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">1</span>))</pre></body></html></div>
</div>
</div>
<div id="are-genetic-differentiation-and-diversity-related" class="section level2">
<h2 class="hasAnchor">
<a href="#are-genetic-differentiation-and-diversity-related" class="anchor"></a>5. Are genetic differentiation and diversity related?</h2>
<p>Would you expect a relationship between genetic diversity and genetic differentiation of individual patches?</p>
<p>Lets examine the correlation between gene diversity (He) and site-specific Fst:</p>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="no">dd.site</span>$<span class="no">He</span>, <span class="no">dd.site</span>$<span class="no">FST.GESTE</span>, <span class="kw">use</span> <span class="kw">=</span> <span class="st">"pairwise.complete"</span>)</pre></body></html></div>
<pre><code>## [1] -0.9567184</code></pre>
<p>There are a number of possible reasons for such a correlation. Can you put forward some hypotheses to explain this relationship? See Lamy et al. (2012) for their interpretation.</p>
</div>
<div id="effect-of-recent-extinction-events" class="section level2">
<h2 class="hasAnchor">
<a href="#effect-of-recent-extinction-events" class="anchor"></a>6. Effect of recent extinction events</h2>
<p>Several patches fell dry between observation years, which is assumed to signify extinction of the local population. Does genetic evidence support this interpretation, i.e., is there genetic evidence of bottlenecks or founder effects in <em>D. depressissimum</em>?</p>
<div id="a-effect-of-patch-extinction-event-temporal-data-set" class="section level3">
<h3 class="hasAnchor">
<a href="#a-effect-of-patch-extinction-event-temporal-data-set" class="anchor"></a>a) Effect of patch extinction event (temporal data set)</h3>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="no">dd.temporal</span> <span class="kw">&lt;-</span> <span class="no">dd.site</span>[<span class="no">dd.site</span>@<span class="kw">data</span>$<span class="no">MultiYear</span><span class="kw">==</span><span class="fl">TRUE</span>,]
<span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="no">dd.temporal</span>@<span class="kw">data</span>[, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Fst.temp"</span>, <span class="st">"APE"</span>, <span class="st">"NLT"</span>, <span class="st">"C"</span>)],
    <span class="kw">use</span><span class="kw">=</span><span class="st">"pairwise.complete"</span>)</pre></body></html></div>
<pre><code>##            Fst.temp         APE         NLT          C
## Fst.temp  1.0000000  0.29379662 -0.10494809 -0.3356817
## APE       0.2937966  1.00000000 -0.07420821 -0.4927664
## NLT      -0.1049481 -0.07420821  1.00000000  0.3103541
## C        -0.3356817 -0.49276638  0.31035409  1.0000000</code></pre>
<p>We can compare a number of competing models using the Akaike Information Criterion (AIC). Models with lower AIC are better (see Week 12).</p>
<div class="sourceCode" id="cb52"><html><body><pre class="r"><span class="no">mod.Fst.temp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">Fst.temp</span>) ~ <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">APE</span>), <span class="kw">data</span><span class="kw">=</span><span class="no">dd.temporal</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">mod.Fst.temp</span>)</pre></body></html></div>
<pre><code>## 
## Call:
## lm(formula = scale(Fst.temp) ~ scale(APE), data = dd.temporal)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -1.0290 -0.5094 -0.3566  0.2140  2.0886 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)
## (Intercept) -0.04099    0.29246  -0.140    0.891
## scale(APE)   0.27599    0.28396   0.972    0.354
## 
## Residual standard error: 1.003 on 10 degrees of freedom
##   (17 observations deleted due to missingness)
## Multiple R-squared:  0.08632,    Adjusted R-squared:  -0.005052 
## F-statistic: 0.9447 on 1 and 10 DF,  p-value: 0.354</code></pre>
<div class="sourceCode" id="cb54"><html><body><pre class="r"><span class="no">mod.Fst.temp.C</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">Fst.temp</span>) ~ <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">APE</span>) + <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">C</span>),
                     <span class="kw">data</span><span class="kw">=</span><span class="no">dd.temporal</span>)
<span class="no">mod.Fst.temp.NLT</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">Fst.temp</span>) ~ <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">APE</span>) + <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">NLT</span>),
                       <span class="kw">data</span><span class="kw">=</span><span class="no">dd.temporal</span>)
<span class="no">mod.Fst.temp.both</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">Fst.temp</span>) ~ <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">APE</span>) + <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">NLT</span>) + <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="no">C</span>),
            <span class="kw">data</span><span class="kw">=</span><span class="no">dd.temporal</span>)
<span class="fu"><a href="https://rdrr.io/r/stats/AIC.html">AIC</a></span>(<span class="no">mod.Fst.temp</span>, <span class="no">mod.Fst.temp.C</span>, <span class="no">mod.Fst.temp.NLT</span>, <span class="no">mod.Fst.temp.both</span>)</pre></body></html></div>
<pre><code>##                   df      AIC
## mod.Fst.temp       3 37.92714
## mod.Fst.temp.C     4 39.27770
## mod.Fst.temp.NLT   4 39.83549
## mod.Fst.temp.both  5 41.27462</code></pre>
<p>The best model includes neither ‘C’ nor ‘NLT’. Note that ‘APE’ is a binary variable, so in essence we’re performing a t-test here.</p>
<div class="sourceCode" id="cb56"><html><body><pre class="r"><span class="no">res.Fst.temp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span>(<span class="no">Fst.temp</span> ~ <span class="no">APE</span>, <span class="kw">data</span><span class="kw">=</span><span class="no">dd.temporal</span>, <span class="kw">alternative</span> <span class="kw">=</span> <span class="st">"less"</span>)
<span class="no">res.Fst.temp</span></pre></body></html></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  Fst.temp by APE
## t = -0.94604, df = 7.893, p-value = 0.1861
## alternative hypothesis: true difference in means is less than 0
## 95 percent confidence interval:
##        -Inf 0.01085372
## sample estimates:
## mean in group FALSE  mean in group TRUE 
##              0.0080              0.0192</code></pre>
</div>
<div id="b-power-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#b-power-analysis" class="anchor"></a>b) Power analysis</h3>
<p>The effect is not statistically significant. Does that mean that we found no effect of apparent population extinctions on temporal Fst? Let’s check effect size. For means, Cohen’s effect size is measured by d (which is measured in units of standard deviations):</p>
<ul>
<li>small effect: d &gt; 0.2 (means at least 0.2 standard deviations apart)</li>
<li>medium effect: d &gt; 0.5</li>
<li>large effect: d &gt; 0.8</li>
</ul>
<p>We can let R calculate effect size for us:</p>
<div class="sourceCode" id="cb58"><html><body><pre class="r"><span class="kw pkg">effsize</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/effsize/man/cohen.d.html">cohen.d</a></span>(<span class="no">Fst.temp</span> ~ <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">APE</span>), <span class="kw">data</span><span class="kw">=</span><span class="no">dd.temporal</span>)</pre></body></html></div>
<pre><code>## 
## Cohen's d
## 
## d estimate: -0.5691221 (medium)
## 95 percent confidence interval:
##      lower      upper 
## -1.8992165  0.7609722</code></pre>
<p>So, we actually found a ‘medium’ effect (more than 0.5 standard deviations difference between group means). Maybe sample size was too small to have sufficient power?</p>
<p>Let’s check sample size:</p>
<div class="sourceCode" id="cb60"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">dd.temporal</span>$<span class="no">APE</span>[!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">dd.temporal</span>$<span class="no">Fst.temp</span>)])</pre></body></html></div>
<pre><code>## 
## FALSE  TRUE 
##     7     5</code></pre>
<p>Ah, that explains a lot. There were only 5 sites with apparent extinction, and 7 without.</p>
<p>Given that sample size, what was the statistical power of our test to detect at least a large effect (d = - 0.8), i.e., be able to reject the null hypothesis if such an effect is present in the population from which we sampled?</p>
<div class="sourceCode" id="cb62"><html><body><pre class="r"><span class="kw pkg">pwr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/pwr/man/pwr.t2n.test.html">pwr.t2n.test</a></span>(<span class="kw">n1</span><span class="kw">=</span><span class="fl">7</span>, <span class="kw">n2</span><span class="kw">=</span><span class="fl">5</span>, <span class="kw">d</span><span class="kw">=</span>-<span class="fl">0.8</span>, <span class="kw">alternative</span> <span class="kw">=</span> <span class="st">"less"</span>)</pre></body></html></div>
<pre><code>## 
##      t test power calculation 
## 
##              n1 = 7
##              n2 = 5
##               d = -0.8
##       sig.level = 0.05
##           power = 0.3552962
##     alternative = less</code></pre>
<p>So the power to detect at least a large effect, if it exists in the population, was only 0.355, way below the 0.8 (or even 0.95) that we would want to see. For a medium effect, the power is even smaller.</p>
</div>
<div id="c-sample-size-calculation" class="section level3">
<h3 class="hasAnchor">
<a href="#c-sample-size-calculation" class="anchor"></a>c) Sample size calculation</h3>
<p>How large a sample would we have needed in each group to achieve a power of 0.8 to detect a large effect? And for a medium effect?</p>
<div class="sourceCode" id="cb64"><html><body><pre class="r"><span class="kw pkg">pwr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/pwr/man/pwr.t.test.html">pwr.t.test</a></span>(<span class="kw">power</span> <span class="kw">=</span> <span class="fl">0.8</span>, <span class="kw">d</span> <span class="kw">=</span> -<span class="fl">0.8</span>, <span class="kw">alternative</span> <span class="kw">=</span> <span class="st">"less"</span>)</pre></body></html></div>
<pre><code>## 
##      Two-sample t test power calculation 
## 
##               n = 20.03277
##               d = -0.8
##       sig.level = 0.05
##           power = 0.8
##     alternative = less
## 
## NOTE: n is number in *each* group</code></pre>
<div class="sourceCode" id="cb66"><html><body><pre class="r"><span class="kw pkg">pwr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/pwr/man/pwr.t.test.html">pwr.t.test</a></span>(<span class="kw">power</span> <span class="kw">=</span> <span class="fl">0.8</span>, <span class="kw">d</span> <span class="kw">=</span> -<span class="fl">0.5</span>, <span class="kw">alternative</span> <span class="kw">=</span> <span class="st">"less"</span>)</pre></body></html></div>
<pre><code>## 
##      Two-sample t test power calculation 
## 
##               n = 50.1508
##               d = -0.5
##       sig.level = 0.05
##           power = 0.8
##     alternative = less
## 
## NOTE: n is number in *each* group</code></pre>
<p>More than 20 sites in each group would have been needed to detect a large effect, or more than 50 per group to detect a medium effect, with a power of 80%.</p>
<p>Hence, these particular results are inconclusive. There was a trend showing a large effect size but power was very low. This aspect of the study should ideally be repeated with a larger sample size before reaching any conclusions.</p>
<p>Note however that using additional evidence (e.g., population assignment tests), Lamy et al. (2012) concluded that extinctions were in fact less common in this system than previously assumed – in many cases of apparent extinction, individuals may still be present but just not detected.</p>
</div>
</div>
<div id="r-exercise-week-4" class="section level2">
<h2 class="hasAnchor">
<a href="#r-exercise-week-4" class="anchor"></a>7. R Exercise Week 4</h2>
<p><strong>Task:</strong> Build on your previous exercises and plot the sites on a map downloaded from the internet. Explore the relationships between Hexp, census population size and percent forest cover within 500 m of the site (forest may act as a barrier for grassland plants).</p>
<p><strong>Hints:</strong></p>
<ol style="list-style-type: lower-alpha">
<li>
<strong>Load packages</strong>: You may want to load the packages <code>dplyr</code> and <code>ggplot2</code>. Alternatively, you can use <code>::</code> to call functions from packages.</li>
<li>
<strong>Import your datasets from Weeks 2 &amp; 3 R Exercises</strong>. Here’s an example for your code, adapt it as needed to import the R objects “Pulsatilla.longlat.rds” (SpatialPointsDataFrame, Week 2) and “H.pop.rds” (Week 3) that you saved previously: <code>Pulsatilla.longlat &lt;- readRDS(paste0(here::here(), "/output/Pulsatilla.longlat.rds"))</code>
</li>
<li>
<strong>Plot sites on map from internet</strong>: To get started, you may use this code (note the names of the variables with the spatial coordinates): <code>myMap &lt;- ggmap::qmplot(x=meanX, y=meanY,  data = as.data.frame(Pulsatilla.longlat), source = "stamen", maptype = "toner-lite")</code> Next, modify code from section 3.d to add labels for all sites and plot the map.</li>
<li>
<p><strong>Combine data</strong>: Use the function <code><a href="https://dplyr.tidyverse.org/reference/join.html">dplyr::left_join</a></code> to add the variables from the dataset <code>H.pop</code> to the data frame in the <span class="citation">@data</span> slot of <code>Pulsatilla.longlat</code>. Notes:</p>
<ul>
<li>This is important, as the order of populations may not be the same in the two datasets.</li>
<li>Remember to check the structure of the datasets (variable names and types) first so that you know which are the ID variables that you can use to match sites.</li>
<li>If the two ID variables are not of the same type (e.g., one if a <code>factor</code>, the other is <code>character</code>), it is best to change the format of one (e.g., with <code>as.character</code>) before doing the left-join.</li>
</ul>
</li>
<li>
<strong>Scatterplot with regression line</strong>: Create a scatterplot of Hexp (y axis) plotted against nIndiv (x axis). Add a regression line and, if possible, label points. You may modify code from section 3.b using the <code>ggplot2</code> package, or use base R functions.<br>
</li>
<li><p><strong>Regression analysis</strong>: Adapt code from section 3.c to perform a regression of <code>Hexp</code> (response variable) on the predictor <code>nIndiv</code>. Create residual plots and inspect them. What is the main issue here?</p></li>
</ol>
<p><strong>Questions:</strong> There is one influential point in the regression analysis:</p>
<ul>
<li>Which site is it?</li>
<li>Where is it located (north / east / south / west)?</li>
<li>What makes it an influential point (large residual, leverage, or both)?</li>
<li>What would happen to the regression line and the R-squared if this point was omitted?</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://sites.utm.utoronto.ca/wagnerlab/">Helene Wagner</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
