<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Week 6: Quantitative Genetics • LandGenCourse</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Week 6: Quantitative Genetics">
<meta property="og:description" content="LandGenCourse">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-168837293-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-168837293-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LandGenCourse</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Week0_BasicR.html">Week 0: Intro to R Programming</a>
    </li>
    <li>
      <a href="../articles/Week0_Graphics.html">Week 0: R Graphics</a>
    </li>
    <li>
      <a href="../articles/Week1_vignette.html">Week 1: Importing Genetic Data</a>
    </li>
    <li>
      <a href="../articles/Week2_vignette.html">Week 2: Spatial Data</a>
    </li>
    <li>
      <a href="../articles/Week2_bonus_vignette.html">Week 2: Bonus Material</a>
    </li>
    <li>
      <a href="../articles/Week3_vignette.html">Week 3: Genetic Diversity</a>
    </li>
    <li>
      <a href="../articles/Week4_vignette.html">Week 4: Metapopulation Genetics</a>
    </li>
    <li>
      <a href="../articles/Week5_vignette.html">Week 5: Spatial Statistics</a>
    </li>
    <li>
      <a href="../articles/Week6_vignette.html">Week 6: Quantitative Genetics</a>
    </li>
    <li>
      <a href="../articles/Week7_vignette.html">Week 7: Spatial Linear Models</a>
    </li>
    <li>
      <a href="../articles/Week8_vignette.html">Week 8: Simulation</a>
    </li>
    <li>
      <a href="../articles/Week8_bonus_vignette.html">Week 8: Bonus Material</a>
    </li>
    <li>
      <a href="../articles/Week9_vignette.html">Week 9: Population Structure</a>
    </li>
    <li>
      <a href="../articles/Week10_vignette.html">Week 10: Landscape Resistance</a>
    </li>
    <li>
      <a href="../articles/Week11_vignette.html">Week 11: Detecting Adaptation</a>
    </li>
    <li>
      <a href="../articles/Week12_vignette.html">Week 12: Model Selection</a>
    </li>
    <li>
      <a href="../articles/Week13_vignette.html">Week 13: Gravity Models</a>
    </li>
    <li>
      <a href="../articles/Week14_vignette.html">Week 14: Contemporary Gene Flow</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../tutorials/00-Intro_RStudio.html">Introduction to RStudio</a>
    </li>
    <li>
      <a href="../tutorials/00-Intro_LandGenCourse.html">Introduction to the LandGenCourse package</a>
    </li>
    <li>
      <a href="../tutorials/00-Intro_RNotebooks.html">Introduction to R Notebooks</a>
    </li>
    <li>
      <a href="../tutorials/00-Version_Control.html">Version Control 101</a>
    </li>
    <li>
      <a href="../tutorials/01-Genetic_Data.html">Week 1 Getting Landscape Genetic Data into R</a>
    </li>
    <li>
      <a href="../tutorials/02-Spatial_Data.html">Week 2 Land Cover Analysis in R</a>
    </li>
    <li>
      <a href="../tutorials/03-Genetic_Diversity.html">Week 3 Genetic Diversity (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/03-Genetic_Diversity_Part2.html">Week 3 Genetic Diversity (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/04-Metapop_Genetics.html">Week 4 Metapopulation Genetics (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/04-Metapop_Genetics_Part2.html">Week 4 Metapopulation Genetics (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/05-Spatial_Stats.html">Week 5 Spatial Statistics</a>
    </li>
    <li>
      <a href="../tutorials/06-Quantitative_Genetics.html">Week 6 Quantitative Genetics (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/06-Quantitative_Genetics_Part2.html">Week 6 Quantitative Genetics (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/07-Spatial_Models.html">Week 7 Spatial Linear Models (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/07-Spatial_Models_Part2.html">Week 7 Spatial Linear Models (Part 2)</a>
    </li>
    <li>
      <a href="../tutorials/08-Modeling.html">Week 8 Simulation Experiments in R (Part 1)</a>
    </li>
    <li>
      <a href="../tutorials/08-Modeling_Part2.html">Week 8 Simulation Experiments in R (Part 2)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Week 6: Quantitative Genetics</h1>
                        <h4 class="author">Andrew Eckert, Katalin Csillery and Helene Wagner</h4>
            
            <h4 class="date">2020-06-08</h4>
      
      
      <div class="hidden name"><code>Week6_vignette.Rmd</code></div>

    </div>

    
    
<div id="overview-of-worked-example" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-of-worked-example" class="anchor"></a>1. Overview of Worked Example</h2>
<div id="a-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#a-goals" class="anchor"></a>a) Goals</h3>
<p><strong>Justification</strong>: Natural selection acts on the heritable component of the total phenotypic variation. Traditionally, the heritable proportion of trait variation is estimated using information on the degree of relatedness between individuals. This approach - extensively employed in breeding and evolutionary quantitative genetics – ignores the molecular underpinnings of traits. An alternative approach is scanning genomes using molecular markers to search for so-called outlier loci (see week 11), where selection invoked a change in allele frequencies that is detectable as unusual differentiation between populations in different environments. This approach – advocated in population genetics - neglects information about the trait variation, but assumes that the markers are in or close to genes that code for adaptive traits. Genetic outliers often hard to interpret because generally we lack information what phenotype they affect and if this phenotype results in fitness differences between populations.</p>
<p><strong>Learning Objectives</strong>: This lab was constructed to give you experience in working with basic quantitative and population genetic analyses to test hypotheses about the presence of local adaptation. Similar to the genomic revolution, phenotyping is also undergoing a revolution thanks to developments e.g. in imagery that allow for a highly automated recording of traits; a task that has been extremely time consuming in the past. These developments will certainly advance the quantitative genetic approach in the future in evolutionary biology let alone breeding. By the end of the laboratory, you should be able to do the following:</p>
<ul>
<li>Construct, fit, and assess linear mixed effects models (LMMs) to estimate genetic values and variance components of a phenotypic trait measured in families, e.g. in a common garden experiments.</li>
<li>Use LMMs to estimate the heritability of a trait.</li>
<li>Test for the presence of unusual trait differentiation among populations relative to differentiation expected in the absence of adaptation</li>
<li>Assess if trait differentiation is correlated with differentiation in environmental variables to try identifying the selective component of the environment.</li>
</ul>
</div>
<div id="b-data-set" class="section level3">
<h3 class="hasAnchor">
<a href="#b-data-set" class="anchor"></a>b) Data set</h3>
<p>The data come from a study of western white pine (Pinus monticola Dougl. ex D. Don) sampled around the Lake Tahoe Basin of California and Nevada. These data consist of 157 trees sampled from 10 populations (n = 9 to 22 trees/population). Within each population, trees were sampled within three plots. For each plot, GPS coordinates were collected (i.e. each plot in each population has its own GPS coordinates) and used to generate a set of 7 environmental variables. From these trees, needle tissue was collected from which total DNA was extracted and genotyped for 164 single nucleotide polymorphisms (SNPs). Seeds were also collected and planted in a common garden. The seedlings (n = 5 to 35/tree) were measured for several phenotypic traits. The phenotypic trait we will be working with today is known as the carbon isotope ratio (<span class="math inline">\(δ^{13}C\)</span>). It is the ratio of two isotopes of carbon (<span class="math inline">\(^{13}C\)</span> and <span class="math inline">\(^{12}C\)</span>) relative to an experimental control, and it is strongly correlated with intrinsic water-use efficiency in plants. Plants need water to live, so it is not a stretch of the imagination to believe that this phenotypic trait has a link with plant fitness.</p>
<p>We will thus have the following data:</p>
<ul>
<li>
<strong>WWP_phenotype_data.txt</strong>: Phenotypic measurements for 5 seedlings per tree made in a common garden.</li>
<li>
<strong>WWP.ecogen</strong>: an ‘ecogen’ object (package ‘EcoGenetics’) with SNP genotypes for all trees sampled in the field, and with environmental data collected from each plot within each population.</li>
</ul>
</div>
<div id="c-required-r-packages" class="section level3">
<h3 class="hasAnchor">
<a href="#c-required-r-packages" class="anchor"></a>c) Required R packages</h3>
<p>Most required packages should have been installed already when you installed ‘LandGenCourse’.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">LandGenCourse</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">lme4</span>)
<span class="co">#require(EcoGenetics)</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">methods</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tibble</span>)
<span class="co">#require(lattice)</span>
<span class="co">#require(MuMIn)</span>
<span class="co">#require(predictmeans)</span>
<span class="co">#require(nlme)</span>
<span class="co">#require(QstFstComp)</span>
<span class="co">#require(car)  </span></pre></body></html></div>
<p>Package ‘ggeffects’ not automatically installed with ‘LandGenCourse’:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="kw">if</span>(!<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">ggeffects</span>)) <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"ggeffects"</span>, <span class="kw">repos</span><span class="kw">=</span><span class="st">'http://cran.us.r-project.org'</span>)</pre></body></html></div>
<pre><code>## Warning in library(package, lib.loc = lib.loc, character.only = TRUE,
## logical.return = TRUE, : there is no package called 'ggeffects'</code></pre>
<pre><code>## 
## The downloaded binary packages are in
##  /var/folders/kz/fv8pzy0d3cl_mp3md294cvbh0000gn/T//RtmpCymydl/downloaded_packages</code></pre>
<p>Source two files with additional functions:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"supplemental_R_functions.R"</span>,
                            <span class="kw">package</span> <span class="kw">=</span> <span class="st">"LandGenCourse"</span>))
<span class="fu"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"panel.cor.r"</span>,
                            <span class="kw">package</span> <span class="kw">=</span> <span class="st">"LandGenCourse"</span>))</pre></body></html></div>
</div>
</div>
<div id="estimate-genetic-and-non-genetic-variance-components-from-a-common-garden-experiment" class="section level2">
<h2 class="hasAnchor">
<a href="#estimate-genetic-and-non-genetic-variance-components-from-a-common-garden-experiment" class="anchor"></a>2. Estimate genetic and non-genetic variance components from a common garden experiment</h2>
<p><strong>Motivation</strong>: A lot of genetics can be carried out without use of any molecular markers. Practitioners of empirical population genetics forget this quite often. A common garden allows us to create a standardized environment in which we minimize the influence of environment on the expression of a particular phenotypic trait. Phenotypic variation can be partitioned to genetic , environmental and residual components based on quantitative genetic theory. Further, we may also test for the presence of interaction between the genetic and environmental variation. The rationale of a common garden is to “standardize” the environment, thus phenotypic variation we observe is mainly due to genetic variation, even though environmental heterogeneity can never be completely eliminated and we still have to additionally control for it, e.g. using a block design (see below).</p>
<p><strong>Goals &amp; Background</strong>: The goal for this part of the laboratory is to construct, fit, and assess LMMs for <span class="math inline">\(δ^{13}C\)</span>. We will be using the data in the file named “WWP_phenotypic_data.txt”. These data are organized in a tab-delimited text file with seedlings grouped by maternal tree (i.e. its mother tree), plot, and population. Also included is an experimental treatment known as “block”. In a common garden, seedlings from the same maternal tree are randomized among blocks to avoid the influence of micro-environmental variation on expression of phenotypic traits.</p>
<div id="a-import-phenytypic-data" class="section level3">
<h3 class="hasAnchor">
<a href="#a-import-phenytypic-data" class="anchor"></a>a) Import phenytypic data</h3>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">phen</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"WWP_phenotype_data.txt"</span>,
                            <span class="kw">package</span> <span class="kw">=</span> <span class="st">"LandGenCourse"</span>), <span class="kw">sep</span> <span class="kw">=</span> <span class="st">"\t"</span>, <span class="kw">header</span> <span class="kw">=</span> <span class="no">T</span>)
<span class="kw pkg">tibble</span><span class="kw ns">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/deprecated.html">as.tibble</a></span>(<span class="no">phen</span>)</pre></body></html></div>
<pre><code>## Warning: `as.tibble()` is deprecated as of tibble 2.0.0.
## Please use `as_tibble()` instead.
## The signature and semantics have changed, see `?as_tibble`.
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_warnings()` to see where this warning was generated.</code></pre>
<pre><code>## # A tibble: 779 x 5
##    population plot  family block  d13c
##    &lt;fct&gt;      &lt;fct&gt;  &lt;int&gt; &lt;int&gt; &lt;dbl&gt;
##  1 blk cyn    BC1       59     5 -30.2
##  2 blk cyn    BC1       59     2 -29.7
##  3 blk cyn    BC1       59     4 -29.6
##  4 blk cyn    BC1       59     3 -29.2
##  5 blk cyn    BC1       59     1 -29.0
##  6 blk cyn    BC1       60     4 -31.2
##  7 blk cyn    BC1       60     3 -30.7
##  8 blk cyn    BC1       60     1 -30.5
##  9 blk cyn    BC1       60     5 -30.1
## 10 blk cyn    BC1       60     2 -29.3
## # … with 769 more rows</code></pre>
<p>Check the variable types: ‘family’ and ‘block’ have been imported as type ‘integer’, and we need to convert them to ‘factor’ first.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">phen</span>$<span class="no">family</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">phen</span>$<span class="no">family</span>)
<span class="no">phen</span>$<span class="no">block</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">phen</span>$<span class="no">block</span>)
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">phen</span>, <span class="no">class</span>)</pre></body></html></div>
<pre><code>## population       plot     family      block       d13c 
##   "factor"   "factor"   "factor"   "factor"  "numeric"</code></pre>
</div>
<div id="b-fit-linear-mixed-effects-models-lmm-to-trait-data-observed-on-families" class="section level3">
<h3 class="hasAnchor">
<a href="#b-fit-linear-mixed-effects-models-lmm-to-trait-data-observed-on-families" class="anchor"></a>b) Fit linear mixed effects models (LMM) to trait data observed on families</h3>
<p>Now,we are ready to fit a series of linear models. Here we will fit four models in total, though the last two are equivalent for our data: - <strong>mod1</strong>: a model with only fixed effects (intercept and block), - <strong>mod2</strong>:a LMM with an fixed effects and a random effect due to family, and - <strong>mod3</strong>:a LMM where the random effect due to family is nested within population. We will thus be ignoring the variable “plot”. - <strong>mod4</strong>:a LMM where the nesting of family within population is not explicitly specified, it is implied by the data.</p>
<p>Note: ‘d13c ~ 1 + block’ is equivalent to ‘d13c ~ block’ (see video, Part 1).</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">mod1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(<span class="no">d13c</span> ~ <span class="no">block</span>, <span class="kw">data</span> <span class="kw">=</span> <span class="no">phen</span>)
<span class="no">mod2</span> <span class="kw">&lt;-</span> <span class="kw pkg">lme4</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmer.html">lmer</a></span>(<span class="no">d13c</span> ~ <span class="fl">1</span> + (<span class="fl">1</span><span class="kw">|</span><span class="no">family</span>) + <span class="no">block</span>,<span class="kw">data</span> <span class="kw">=</span> <span class="no">phen</span>, <span class="kw">REML</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">mod3</span> <span class="kw">&lt;-</span> <span class="kw pkg">lme4</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmer.html">lmer</a></span>(<span class="no">d13c</span> ~ <span class="fl">1</span> + (<span class="fl">1</span><span class="kw">|</span><span class="no">population</span>/<span class="no">family</span>) + <span class="no">block</span>,
                   <span class="kw">data</span> <span class="kw">=</span> <span class="no">phen</span>, <span class="kw">REML</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">mod4</span> <span class="kw">&lt;-</span> <span class="kw pkg">lme4</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmer.html">lmer</a></span>(<span class="no">d13c</span> ~ <span class="fl">1</span> + (<span class="fl">1</span><span class="kw">|</span><span class="no">population</span>) + (<span class="fl">1</span><span class="kw">|</span><span class="no">family</span>) + <span class="no">block</span>,
                   <span class="kw">data</span> <span class="kw">=</span> <span class="no">phen</span>, <span class="kw">REML</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p>Models ‘mod3’ and ‘mod4’ produce identical results, as long as each family has its own unique ID. In the following, we will use ‘mod4’ as it simplifies interpretation (as commonly used in quantitative genetics).</p>
<p>Notice that we are using REML=TRUE, which stands for restricted maximum likelihood. This method is advised for quantitative genetic analysis, because it provides unbiased estimates of the variance components. The other approach would be to use ML or maximum likelihood, however, it generally underestimates the residual variance, thus leading to inflated estimates of the family variance and thus the heritability. ML would be needed e.g. to test fixed effects.</p>
</div>
<div id="c-compare-model-fit" class="section level3">
<h3 class="hasAnchor">
<a href="#c-compare-model-fit" class="anchor"></a>c) Compare model fit</h3>
<p>Now, let’s explore which model best fits our data. To do this we will use the Akaike Information Criterion (AIC). This statistic scores model fit while giving a penalty for the number of parameters in a model. The model with the lowest score is the preferred model.</p>
<p>How do we ensure that the AIC values of the different models are comparable?</p>
<ul>
<li>For ‘mod1’ that was fitted with ‘lm’, we use the function ‘AIC’.</li>
<li>For the other models that were fitted with ‘lmer’, we use ‘extractAIC’.</li>
</ul>
<p>For model comparison purposes, we have to use the ML fit, as AIC is no longer valid for REML. The function ‘extractAIC’ refits the models with ‘REML=FALSE’ to obtain AIC values that are comparable between models with different fixed effects (though this does not apply here because the fixed effects were the same), or between models fitted with ‘lm’ and ‘lmer’. It returns two values, the first is the ‘equivalent degrees of freedom’ of the fitted model, and the second is the ‘AIC’ value. Here we only extract the second value.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">aic_vals</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/AIC.html">AIC</a></span>(<span class="no">mod1</span>), <span class="fu"><a href="https://rdrr.io/r/stats/extractAIC.html">extractAIC</a></span>(<span class="no">mod2</span>)[<span class="fl">2</span>], <span class="fu"><a href="https://rdrr.io/r/stats/extractAIC.html">extractAIC</a></span>(<span class="no">mod3</span>)[<span class="fl">2</span>],
              <span class="fu"><a href="https://rdrr.io/r/stats/extractAIC.html">extractAIC</a></span>(<span class="no">mod4</span>)[<span class="fl">2</span>])
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">aic_vals</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"mod1"</span>,<span class="st">"mod2"</span>,<span class="st">"mod3"</span>, <span class="st">"mod4"</span>)
<span class="no">aic_vals</span></pre></body></html></div>
<pre><code>##     mod1     mod2     mod3     mod4 
## 2120.987 2080.266 2051.207 2051.207</code></pre>
<p>Hence, model ‘mod3’ (and its equivalent ‘mod4’) appears to be the best model. This suggests that there are important differences among populations, and among trees within populations.</p>
<p>We will learn more about model selection later in the course (Week 12).</p>
</div>
<div id="d-check-model-validity" class="section level3">
<h3 class="hasAnchor">
<a href="#d-check-model-validity" class="anchor"></a>d) Check model validity</h3>
<p>Is the model ‘mod4’ valid? Let’s check the residuals. The function ‘residplot’ from package ‘predictmeans’ produces four diagnostic plots for a model fitted with ‘lmer’. The option ‘level’ indicates which random factor should be plotted (here: 1 = family, 2 = population). This function uses the ‘conditional’ residuals, which represent ‘observed - fitted(fixed) - fitted(random)’.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="kw pkg">predictmeans</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/predictmeans/man/residplot.html">residplot</a></span>(<span class="no">mod4</span>, <span class="kw">group</span><span class="kw">=</span><span class="st">"population"</span>, <span class="kw">level</span><span class="kw">=</span><span class="fl">1</span>)</pre></body></html></div>
<p>Hint: You may need to use the arrow symbols to scroll through the plots in the ‘Plot’ tab in RStudio.</p>
<p>The plots are:</p>
<ul>
<li>A normal probability plot of the random effect ‘family’: points should follow line.</li>
<li>A normal probability plot of (conditional) residuals: points should follow line.</li>
<li>A scatterplot of the (conditional) residuals against fitted values: the plot should not ‘thicken’ (which would indicate differences in variances between groups such as blocks = heteroscedasticity). Due to the option ’group = “population”, residuals are colored by population.</li>
<li>A plot of fitted values against observed: the scatterplot shows the overall model fit, where points would fall on a line if the model explained 100% of the variation in the response (you can ignore the dashed line).</li>
</ul>
<p>In addition, we may want to look for outliers and influential points. For this, we create two additional plots. Here we use marginal residuals, which are calculated without accounting for random effects: ‘observed - fitted(fixed)’.</p>
<ul>
<li>A plot of marginal residuals against the fixed effect ‘block’: there should not be any large outliers.<br>
</li>
<li>A plot of Cook’s distances, where D &gt; 1 indicates influential points (more relevant when using quantitative predictors (covariates). Note: this may take a while to calculate.</li>
</ul>
<p>First we calculate the marginal residuals by predicting values with fixed effects only (‘re.form=NA’) and subtract these fitted values from the observed values.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">marginal.residuals</span> <span class="kw">&lt;-</span> <span class="no">mod4</span>@<span class="kw">frame</span>$<span class="no">d13c</span> - <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(<span class="no">mod4</span>, <span class="kw">re.form</span><span class="kw">=</span><span class="fl">NA</span>)
<span class="fu"><a href="https://rdrr.io/pkg/ggeffects/man/plot.html">plot</a></span>(<span class="no">mod4</span>@<span class="kw">frame</span>$<span class="no">block</span>, <span class="no">marginal.residuals</span>)</pre></body></html></div>
<p><img src="Week6_vignette_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co">#predictmeans::CookD(mod4)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ggeffects/man/plot.html">plot</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/influence.measures.html">cooks.distance</a></span>(<span class="no">mod4</span>))</pre></body></html></div>
<p><img src="Week6_vignette_files/figure-html/unnamed-chunk-9-2.png" width="700"></p>
<p>The residual plots did not indicate any major problem. Hence we can proceed with interpreting the results.</p>
</div>
<div id="e-estimate-variance-components" class="section level3">
<h3 class="hasAnchor">
<a href="#e-estimate-variance-components" class="anchor"></a>e) Estimate variance components</h3>
<p>How much of the variation in <span class="math inline">\(d^{13}C\)</span>’ is explained by the fixed and random effects? In models fitted with ‘lm’, we use the <span class="math inline">\(R^2\)</span> statistics to describe the total variance in the response explained by all the predictors in the model. In a mixed effects modeling context, there is no longer an obvious definition for <span class="math inline">\(R^2\)</span>. Two approximate <span class="math inline">\(R^2\)</span> statistics have been implemented in the function ‘r.squaredGLMM’ from the package ‘MuMIn’ (which stands for multi-model inference):</p>
<ul>
<li>R2m: marginal <span class="math inline">\(R^2\)</span>. This is the variance explained by fixed factors.</li>
<li>R2c: conditional <span class="math inline">\(R^2\)</span>. This is interpreted as the variance explained by both fixed and random factors, i.e., the entire model.</li>
</ul>
<p>In both statistics, the denominator is a sum of the fixed-effects variance, the random effect variance, and the residual variance.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="kw pkg">MuMIn</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/MuMIn/man/r.squaredGLMM.html">r.squaredGLMM</a></span>(<span class="no">mod4</span>)</pre></body></html></div>
<pre><code>## Warning: 'r.squaredGLMM' now calculates a revised statistic. See the help page.</code></pre>
<pre><code>##             R2m       R2c
## [1,] 0.02466321 0.2297893</code></pre>
<p>The fixed effect ‘block’ had a small effect of about 2.5%. The total model explained about 23%, most of which was due to the random effects.</p>
<p>How important are the random factors ‘population’ and ‘family’? The summary for ‘mod4’ lists the variance components under ‘Random effects’.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">mod4</span>)</pre></body></html></div>
<pre><code>## Linear mixed model fit by REML ['lmerMod']
## Formula: d13c ~ 1 + (1 | population) + (1 | family) + block
##    Data: phen
## 
## REML criterion at convergence: 2050.4
## 
## Scaled residuals: 
##     Min      1Q  Median      3Q     Max 
## -2.5436 -0.7485  0.0151  0.6028  3.7867 
## 
## Random effects:
##  Groups     Name        Variance Std.Dev.
##  family     (Intercept) 0.08164  0.2857  
##  population (Intercept) 0.10859  0.3295  
##  Residual               0.71429  0.8452  
## Number of obs: 779, groups:  family, 157; population, 10
## 
## Fixed effects:
##              Estimate Std. Error  t value
## (Intercept) -30.62635    0.12666 -241.792
## block2       -0.13833    0.09667   -1.431
## block3       -0.35071    0.09520   -3.684
## block4       -0.10060    0.09538   -1.055
## block5       -0.39443    0.09651   -4.087
## 
## Correlation of Fixed Effects:
##        (Intr) block2 block3 block4
## block2 -0.374                     
## block3 -0.379  0.498              
## block4 -0.378  0.497  0.505       
## block5 -0.374  0.492  0.499  0.496</code></pre>
<p>Here we extract these variance components from the summary and divide by their sum.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">fam.var</span> <span class="kw">&lt;-</span> <span class="kw pkg">nlme</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/nlme/man/VarCorr.html">VarCorr</a></span>(<span class="no">mod4</span>)$<span class="st">"family"</span>[<span class="fl">1</span>]
<span class="no">prov.var</span> <span class="kw">&lt;-</span> <span class="kw pkg">nlme</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/nlme/man/VarCorr.html">VarCorr</a></span>(<span class="no">mod4</span>)$<span class="st">"population"</span>[<span class="fl">1</span>]
<span class="no">res.var</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">mod4</span>)$<span class="no">sigma</span>^<span class="fl">2</span>

<span class="no">Components</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">fam.var</span>, <span class="no">prov.var</span>, <span class="no">res.var</span>)
<span class="no">Components</span> / <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">Components</span>)</pre></body></html></div>
<pre><code>##      fam.var  prov.var   res.var
## 1 0.09026004 0.1200531 0.7896869</code></pre>
<p>Compared to Cohen’s (1988) effect sizes, population and family each had a medium-size effect (&gt; 9% variance explained) on d13c values.</p>
</div>
<div id="f-significance-testing" class="section level3">
<h3 class="hasAnchor">
<a href="#f-significance-testing" class="anchor"></a>f) Significance testing</h3>
<p>In quantiative genetics, we are more interested in estimating variance components and effects than hypothesis testing. Let’s do some testing anyways to see how it works.</p>
<p>The confusing part here is that we need to fit the model differentely to test fixed and random effects. For random effects, we can use model ‘mod4’ that was fitted with ‘REML=TRUE’.</p>
<p>The simplest way to test the significance of a random effect is to calculate the model with and without it and use ‘anova()’ to test whether the more complex model (listed first) explains significantly more than the simpler model. This implements a likelihood ratio test (LR). By default, when used for a model fitted with ‘lmer’, ‘anova’ will refit the models with ML. Here we use the option’refit=FALSE’ to prevent this.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="co">#mod.noPop &lt;- update(mod4, .~. -(1 | population))</span>
<span class="no">mod.noPop</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmer.html">lmer</a></span>(<span class="no">d13c</span> ~ <span class="fl">1</span>  + (<span class="fl">1</span> <span class="kw">|</span> <span class="no">family</span>) + <span class="no">block</span>, <span class="kw">data</span><span class="kw">=</span><span class="no">phen</span>, <span class="kw">REML</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="no">mod.noFam</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmer.html">lmer</a></span>(<span class="no">d13c</span> ~ <span class="fl">1</span>  + (<span class="fl">1</span> <span class="kw">|</span> <span class="no">population</span>) + <span class="no">block</span>, <span class="kw">data</span><span class="kw">=</span><span class="no">phen</span>, <span class="kw">REML</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span>(<span class="no">mod4</span>, <span class="no">mod.noPop</span>, <span class="kw">refit</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
<pre><code>## Data: phen
## Models:
## mod.noPop: d13c ~ 1 + (1 | family) + block
## mod4: d13c ~ 1 + (1 | population) + (1 | family) + block
##           npar    AIC    BIC  logLik deviance  Chisq Df Pr(&gt;Chisq)    
## mod.noPop    7 2097.2 2129.8 -1041.6   2083.2                         
## mod4         8 2066.4 2103.7 -1025.2   2050.4 32.819  1  1.012e-08 ***
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span>(<span class="no">mod4</span>, <span class="no">mod.noFam</span>, <span class="kw">refit</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
<pre><code>## Data: phen
## Models:
## mod.noFam: d13c ~ 1 + (1 | population) + block
## mod4: d13c ~ 1 + (1 | population) + (1 | family) + block
##           npar    AIC    BIC  logLik deviance  Chisq Df Pr(&gt;Chisq)    
## mod.noFam    7 2077.4 2110.0 -1031.7   2063.4                         
## mod4         8 2066.4 2103.7 -1025.2   2050.4 12.996  1  0.0003122 ***
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
<p>We can do a similar test for the fixed effecs. Here we have only one effect.</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="no">mod.noBlock</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmer.html">lmer</a></span>(<span class="no">d13c</span> ~ <span class="fl">1</span>  + (<span class="fl">1</span> <span class="kw">|</span> <span class="no">population</span>) + (<span class="fl">1</span> <span class="kw">|</span> <span class="no">family</span>), <span class="kw">data</span><span class="kw">=</span><span class="no">phen</span>, <span class="kw">REML</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="no">mod.noFam</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmer.html">lmer</a></span>(<span class="no">d13c</span> ~ <span class="fl">1</span>  + (<span class="fl">1</span> <span class="kw">|</span> <span class="no">population</span>) + <span class="no">block</span>, <span class="kw">data</span><span class="kw">=</span><span class="no">phen</span>, <span class="kw">REML</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span>(<span class="no">mod4</span>, <span class="no">mod.noBlock</span>, <span class="kw">refit</span><span class="kw">=</span><span class="fl">TRUE</span>, <span class="kw">REML</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
<pre><code>## refitting model(s) with ML (instead of REML)</code></pre>
<pre><code>## Data: phen
## Models:
## mod.noBlock: d13c ~ 1 + (1 | population) + (1 | family)
## mod4: d13c ~ 1 + (1 | population) + (1 | family) + block
##             npar    AIC    BIC  logLik deviance  Chisq Df Pr(&gt;Chisq)    
## mod.noBlock    4 2067.7 2086.3 -1029.8   2059.7                         
## mod4           8 2051.2 2088.5 -1017.6   2035.2 24.454  4  6.475e-05 ***
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
<p>If we have several fixed effects, it may be more convenient to use function ‘Anova’ from the ‘car’ package to perform a Wald chi-square test. However, the model must be fitted with ML. Choose between type II and type II sums of squares (see video).</p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="no">mod4.ML</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmer.html">lmer</a></span>(<span class="no">d13c</span> ~ <span class="fl">1</span>  + (<span class="fl">1</span> <span class="kw">|</span> <span class="no">population</span>) + (<span class="fl">1</span> <span class="kw">|</span> <span class="no">family</span>) + <span class="no">block</span>,
                <span class="kw">data</span><span class="kw">=</span><span class="no">phen</span>, <span class="kw">REML</span><span class="kw">=</span><span class="fl">FALSE</span>)
<span class="kw pkg">car</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/car/man/Anova.html">Anova</a></span>(<span class="no">mod4.ML</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"II"</span>, <span class="kw">test.statistic</span><span class="kw">=</span><span class="st">"Chisq"</span>)</pre></body></html></div>
<pre><code>## Registered S3 methods overwritten by 'car':
##   method                          from
##   influence.merMod                lme4
##   cooks.distance.influence.merMod lme4
##   dfbeta.influence.merMod         lme4
##   dfbetas.influence.merMod        lme4</code></pre>
<pre><code>## Analysis of Deviance Table (Type II Wald chisquare tests)
## 
## Response: d13c
##        Chisq Df Pr(&gt;Chisq)    
## block 24.905  4  5.257e-05 ***
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<div id="estimate-trait-heritability" class="section level2">
<h2 class="hasAnchor">
<a href="#estimate-trait-heritability" class="anchor"></a>3. Estimate trait heritability</h2>
<p><strong>Motivation</strong>: Now that we have learned how to estimate genetic values for <span class="math inline">\(δ^{13}C\)</span>, let’s learn how to estimate what fraction of the total variation in trait values is due to genetic effects. More precisely, we shall estimate the heritable proportion of the trait variation. Not all genetic effects are transmitted from one generation to the next, but the so-called additive genetic effects (see Week 6 lecture). Here we have data on half-siblings because seeds come from the same mother tree, but their father is most likely different because our species is wind pollinated. In such family structure, the estimation of the additive genetic variance is straightforward because these analyses provide key information about whether or not local adaptation should even be considered. Remember that local adaptation is about genetically determined phenotypes that vary across environments in responses to differing selective pressures. This step allows us to assess how genetic variation for a phenotypic trait is distributed across the landscape.</p>
<p><strong>Goals &amp; Background</strong>: The goal for this part of the laboratory is to estimate heritability, trait differentiation, and correlation with environment for trait values determined in Part 1. To do this, we will be using the output from the previous part of the laboratory and the environmental data contained in the file named “WWP_environmental_data.txt”. As with the phenotype file this is a tab-delimited text file.</p>
<div id="a-estimate-heritability" class="section level3">
<h3 class="hasAnchor">
<a href="#a-estimate-heritability" class="anchor"></a>a) Estimate heritability</h3>
<p>Let’s start with estimating the heritability of <span class="math inline">\(δ^{13}C\)</span>. If you remember from your undergraduate evolution course, heritability refers generally to the proportion of phenotypic variance due to genetic variance. It comes in at least two different versions. The version we are interested in is narrow-sense heritability (<span class="math inline">\(h^2\)</span>), which is defined as the ratio of additive genetic variance to total phenotypic variance:</p>
<p><span class="math display">\[h^{2} = \frac{\sigma^{2}_{additive}}{\sigma^{2}_{total}}\]</span> We need to extract the variance components from ‘mod3’ for all model terms. We do this visually by printing mod3 to screen or using a set of functions applied to ‘mod3’. Here, we will do both.</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">mod4</span></pre></body></html></div>
<pre><code>## Linear mixed model fit by REML ['lmerMod']
## Formula: d13c ~ 1 + (1 | population) + (1 | family) + block
##    Data: phen
## REML criterion at convergence: 2050.381
## Random effects:
##  Groups     Name        Std.Dev.
##  family     (Intercept) 0.2857  
##  population (Intercept) 0.3295  
##  Residual               0.8452  
## Number of obs: 779, groups:  family, 157; population, 10
## Fixed Effects:
## (Intercept)       block2       block3       block4       block5  
##    -30.6263      -0.1383      -0.3507      -0.1006      -0.3944</code></pre>
<p>Using the results from above, let’s calculate <span class="math inline">\(h^2\)</span>. If we assume that the seedlings from each maternal tree are half-siblings (i.e. same mom, but each with a different father) then <span class="math inline">\(σ^2_A = 4 σ^2_{family}\)</span> (so variance due to family). If the seedlings were all full-siblings, then the 4 would be replaced with 2. Let’s assume half-siblings. We can then do the following:</p>
<p>Copying the values visually from the output above (note that we need to square the standard deviations to get the variances):</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="no">add_var</span> <span class="kw">&lt;-</span> <span class="fl">4</span>*(<span class="fl">0.2857</span>^<span class="fl">2</span>)
<span class="no">total_wp_var</span> <span class="kw">&lt;-</span> (<span class="fl">0.2857</span>^<span class="fl">2</span>) + (<span class="fl">0.3295</span>^<span class="fl">2</span>) + (<span class="fl">0.8452</span>^<span class="fl">2</span>)
<span class="no">h2</span> <span class="kw">&lt;-</span> <span class="no">add_var</span>/<span class="no">total_wp_var</span>
<span class="no">h2</span></pre></body></html></div>
<pre><code>## [1] 0.3609476</code></pre>
<p>And the same using the variance components from above:</p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="no">one.over.relatedness</span> <span class="kw">&lt;-</span> <span class="fl">1</span>/<span class="fl">0.25</span>
<span class="no">h2</span> <span class="kw">&lt;-</span> (<span class="no">one.over.relatedness</span>*<span class="no">Components</span>$<span class="no">fam.var</span>) / <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">Components</span>)
<span class="no">h2</span></pre></body></html></div>
<pre><code>## [1] 0.3610402</code></pre>
<p><strong>Question:</strong> Inspect your value of <span class="math inline">\(h^2\)</span>. What does it mean?</p>
<p>We have generated a point estimate for <span class="math inline">\(h^2\)</span>. It represents the average <span class="math inline">\(h^2\)</span> across populations after removing the genetic effects due to population differences. To take this further and also get a SE (so that you can construct a confidence interval), you could check out this R package: <a href="https://bioconductor.org/packages/release/bioc/html/GeneticsPed.html" class="uri">https://bioconductor.org/packages/release/bioc/html/GeneticsPed.html</a></p>
</div>
<div id="b-estimate-trait-differentiation" class="section level3">
<h3 class="hasAnchor">
<a href="#b-estimate-trait-differentiation" class="anchor"></a>b) Estimate trait differentiation</h3>
<p>Great, we have shown that within population genetic variation is statistically greater than zero. What about among population genetic variation? Let’s get to that right now. To measure among population genetic variation we will use a statistic known as <span class="math inline">\(Q_{ST}\)</span>. It is similar in concept to <span class="math inline">\(F_{ST}\)</span> from population genetics. To estimate <span class="math inline">\(Q_{ST}\)</span>, we will use our LMM output again. If we assume that all seedlings are again half-siblings, then:</p>
<p><span class="math display">\[Q_{ST} = \frac{\sigma^{2}_{population}}
{\sigma^{2}_{population}+8\sigma^{2}_{family}}\]</span></p>
<p>Again, we can do this by copying the values visually from the output:</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="no">num_qst</span> <span class="kw">&lt;-</span> <span class="fl">0.3295</span>^<span class="fl">2</span>
<span class="no">dem_qst</span> <span class="kw">&lt;-</span> <span class="fl">0.3295</span>^<span class="fl">2</span> + (<span class="fl">8</span>*(<span class="fl">0.2857</span>^<span class="fl">2</span>))
<span class="no">qst</span> <span class="kw">&lt;-</span> <span class="no">num_qst</span>/<span class="no">dem_qst</span>
<span class="no">qst</span></pre></body></html></div>
<pre><code>## [1] 0.1425618</code></pre>
<p>Or using the variance components from above:</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="no">num_qst</span> <span class="kw">&lt;-</span> <span class="no">Components</span>$<span class="no">prov.var</span>
<span class="no">dem_qst</span> <span class="kw">&lt;-</span> <span class="no">Components</span>$<span class="no">prov.var</span> + (<span class="fl">8</span>*<span class="no">Components</span>$<span class="no">fam.var</span>)
<span class="no">qst</span> <span class="kw">&lt;-</span> <span class="no">num_qst</span>/<span class="no">dem_qst</span>
<span class="no">qst</span></pre></body></html></div>
<pre><code>## [1] 0.1425583</code></pre>
<p><strong>Question:</strong> Inspect your ‘qst’ value:</p>
<ul>
<li>What does it mean?</li>
<li>Look at the quantities in the equation above, what is the denominator equal to? Is it the total phenotypic variance or the total genetic variance?</li>
</ul>
</div>
</div>
<div id="compare-q_st-to-f_st-measured-from-snp-data" class="section level2">
<h2 class="hasAnchor">
<a href="#compare-q_st-to-f_st-measured-from-snp-data" class="anchor"></a>4. Compare <span class="math inline">\(Q_{st}\)</span> to <span class="math inline">\(F_{st}\)</span> measured from SNP data</h2>
<p>If you want to learn more about analyzing SNP data in R, here are some resources:</p>
<ul>
<li>
<strong>adegenet SNP tutorial</strong>: <a href="http://adegenet.r-forge.r-project.org/files/tutorial-genomics.pdf" class="uri">http://adegenet.r-forge.r-project.org/files/tutorial-genomics.pdf</a>
</li>
<li>
<strong>popgen.nescent.org</strong> (click on ‘Users’ tab): <a href="http://popgen.nescent.org/" class="uri">http://popgen.nescent.org/</a>
</li>
</ul>
<div id="a-import-ecogen-object" class="section level3">
<h3 class="hasAnchor">
<a href="#a-import-ecogen-object" class="anchor"></a>a) Import ‘ecogen’ object</h3>
<p>Load the data set ‘WWP.ecogen’.</p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">WWP.ecogen</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"LandGenCourse"</span>)</pre></body></html></div>
<p>This ‘ecogen’ object has information in the following slots:</p>
<ul>
<li>
<strong>XY</strong>: data frame with spatial coordinates (here: longitude, latitude)</li>
<li>
<strong>P</strong>: data frame with phenotypic traits (here: intercept, family effects and population effects from mod4)</li>
<li>
<strong>G</strong>: data frame with genotypic data (here: 164 SNPs)</li>
<li>
<strong>A</strong>: generated automatically from G: matrix of allele counts (codominant markers only)</li>
<li>
<strong>E</strong>: data frame with environmental data (here: 7 bioclimatic site variables)</li>
<li>
<strong>S</strong>: data frame with structure (hierarchical sampling levels)</li>
</ul>
</div>
<div id="b-compare-q_st-to-f_st" class="section level3">
<h3 class="hasAnchor">
<a href="#b-compare-q_st-to-f_st" class="anchor"></a>b) Compare <span class="math inline">\(Q_{ST}\)</span> to <span class="math inline">\(F_{ST}\)</span>
</h3>
<p>See ‘Week 6 Bonus Material’ for calculating Fst from the SNP data.</p>
<p>Here, we will use the ‘QstFstComp’ library to formally test whether or not <span class="math inline">\(Q_{ST} &gt; F_{ST}\)</span> for <span class="math inline">\(δ^{13}C\)</span>. The function ‘QstFstComp’ from the package with the same name calculates both <span class="math inline">\(Q_{ST}\)</span> and <span class="math inline">\(F_{ST}\)</span> itself, then tests whethere they are different from each other.</p>
<p>Note: as this is a permutation test, if you run it several times, the results may change slightly from run to run.</p>
<p>First, however, we need to re-format the data to the format that ‘QstFstComp’ expects. This is easy with the function ‘ecogen2hierfstat’.</p>
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="no">WWP.hierfstat</span> <span class="kw">&lt;-</span> <span class="kw pkg">EcoGenetics</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/EcoGenetics/man/ecogen2hierfstat.html">ecogen2hierfstat</a></span>(<span class="no">WWP.ecogen</span>, <span class="kw">pop</span><span class="kw">=</span><span class="st">'population'</span>,
                                  <span class="kw">to_numeric</span><span class="kw">=</span><span class="fl">TRUE</span>, <span class="kw">nout</span><span class="kw">=</span><span class="fl">1</span>)
<span class="no">phen_mod</span> <span class="kw">&lt;-</span> <span class="no">phen</span>[,-<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,<span class="fl">4</span>)]
<span class="no">QstFst_out</span> <span class="kw">&lt;-</span> <span class="kw pkg">QstFstComp</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/QstFstComp/man/QstFstComp.html">QstFstComp</a></span>(<span class="kw">fst.dat</span> <span class="kw">=</span> <span class="no">WWP.hierfstat</span>,
                                     <span class="kw">qst.dat</span> <span class="kw">=</span> <span class="no">phen_mod</span>,
                                     <span class="kw">numpops</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nlevels.html">nlevels</a></span>(<span class="no">WWP</span>@<span class="kw">S</span>$<span class="no">population</span>),
                                     <span class="kw">nsim</span> <span class="kw">=</span> <span class="fl">10000</span>,
                                     <span class="kw">breeding.design</span> <span class="kw">=</span> <span class="st">"half.sib.dam"</span>,
                                     <span class="kw">dam.offspring.relatedness</span> <span class="kw">=</span> <span class="fl">0.25</span>,
                                     <span class="kw">output</span> <span class="kw">=</span> <span class="st">"concise_nowrite"</span>)</pre></body></html></div>
<pre><code>## Registered S3 method overwritten by 'spdep':
##   method   from
##   plot.mst ape</code></pre>
<pre><code>## [1] "No output file of Q minus F values written."</code></pre>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="no">QstFst_out</span></pre></body></html></div>
<pre><code>## [[1]]
##            Calculated Qst-Fst  Lower Bound crit. value.2.5% 
##                    0.11917632                   -0.02735188 
## Upper bound crit. value.97.5% 
##                    0.06420595 
## 
## [[2]]
## [1] "Qst-Fst values output to file: /Users/helene/Desktop/R_Github_projects/LandGenCourse/vignettes/QminusFvalues_2020-06-08_22-23-14.txt"
## 
## [[3]]
## Lower one-tailed p value Upper one-tailed p value       Two-tailed p value 
##                    0.995                    0.005                    0.010 
## 
## [[4]]
##        Estimated Fst  Lower Bound CI.2.5% Upper bound CI.97.5% 
##           0.01419871           0.01110717           0.01741976 
## 
## [[5]]
##        Estimated Qst  Lower Bound CI.2.5% Upper bound CI.97.5% 
##           0.13337503           0.02725072           0.36234904 
## 
## [[6]]
##                   Va  Lower bound CI.2.5% Upper bound CI.97.5% 
##            0.3180848            0.1065291            0.5436353</code></pre>
<p>The output contains the following elements:</p>
<ul>
<li>‘[[1]]’: The calculated difference between Qst and Fst, with 95% critical values of the distribution of the distribution under the null hypothesis of no difference.</li>
<li>‘[[2]]’: an output file name (though we suppressed the writing of this file with the option ‘output=“consise_nowrite”)’</li>
<li>‘[[3]]’: p-values for a hypothesis test with H0: <span class="math inline">\(Q_{ST} = F_{ST}\)</span>, with three different alternatives (‘less’, ‘greater’, ‘two.sided’)</li>
<li>‘[[4]]’: the Fst value estimated from the data with 95% confidence intervals.</li>
<li>‘[[5]]’: the Qst value estimated from the data with 95% confidence intervals.</li>
<li>‘[[6]]’: the additive genetic variance for the trait with 95% confidence intervals.</li>
</ul>
<p>Note: the values are somewhat different from what we calculated, most likely because the function ‘QstFstComp’ did not account for the block effect.</p>
<p><strong>Questions:</strong> Inspect the first and third elements of the list ‘QstFst_out’:</p>
<ul>
<li>For the observed values, is <span class="math inline">\(Q_{ST} &gt; F_{ST}\)</span>?</li>
<li>Which alternative is most appropriate here?</li>
<li>Can the null hypothesis be rejected with alpha = 0.05?</li>
<li>What does this mean biologically?</li>
</ul>
</div>
</div>
<div id="assess-correlation-between-trait-and-environment" class="section level2">
<h2 class="hasAnchor">
<a href="#assess-correlation-between-trait-and-environment" class="anchor"></a>5. Assess correlation between trait and environment</h2>
<p><strong>Motivation</strong>: Similar to the fixed effects, estimates of the random effects can be also extracted from a mixed effects model and used in further analysis. Such estimates are called Best Linear Unbiased Predictions or BLUPs. They are technically predictions because when you fit a mixed effects model, initially only the mean and variance of the random effects are estimated. Then, the random effects or BLUPs for a given experimental unit (e.g., population or family) can be “predicted” from the estimated mean and variance (and the data).</p>
<p><strong>Goals and background</strong>: Here we will extract block, population and family effects from the model. Then we will correlate the plot-level bioclimatic data with the population effects to see if population differences in seedling d13C trait can be explained by the environment at the site where the mother tree stands. The underlying idea is to look if selection has made the population different.</p>
<div id="a-describe-and-extract-effects" class="section level3">
<h3 class="hasAnchor">
<a href="#a-describe-and-extract-effects" class="anchor"></a>a) Describe and extract effects</h3>
<p>Which blocks, populations or trees had higher mean d13c values than others? To answer this question, we extract and plot the fixed and random effects. We’ll start with the fixed effect for ‘block’.</p>
<p>The effects are not simply the mean of all seedlings from the same block, but the marginal effect of ‘block’, i.e., estimates of the mean per block after accounting for all other effects (i.e., keeping population and family constant).</p>
<p>There are two ways to extract the block effects. The first is with function ‘fixef’. This treats the first level of ‘block’ as a reference level and uses it to estimate the global intercept, then expresses the effect of the other blocks as deviations from the first block.</p>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="kw pkg">lme4</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/lme4/man/fixef.html">fixef</a></span>(<span class="no">mod4</span>)</pre></body></html></div>
<pre><code>## (Intercept)      block2      block3      block4      block5 
## -30.6263478  -0.1383311  -0.3507053  -0.1006031  -0.3944314</code></pre>
<p>A more convenient way to extract the effects is with function ‘ggeffect’ of the package ‘ggeffects’.</p>
<p>Note: The package ‘ggeffects’ is not automatically imported with ‘LandGenCourse’ to avoid incompatibility issues. You can install by runing ‘install.packages(“ggeffects”)’ below without ‘#’.</p>
<div class="sourceCode" id="cb52"><html><body><pre class="r"><span class="co"># install.packages("ggeffects")</span>

<span class="kw pkg">ggeffects</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggeffects/man/ggpredict.html">ggeffect</a></span>(<span class="no">mod4</span>, <span class="kw">terms</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"block"</span>))</pre></body></html></div>
<pre><code>## 
## # Predicted values of d13c
## # x = block
## 
## x | Predicted |   SE |           95% CI
## ---------------------------------------
## 1 |    -30.63 | 0.13 | [-30.87, -30.38]
## 2 |    -30.76 | 0.13 | [-31.01, -30.51]
## 3 |    -30.98 | 0.13 | [-31.23, -30.73]
## 4 |    -30.73 | 0.13 | [-30.98, -30.48]
## 5 |    -31.02 | 0.13 | [-31.27, -30.77]</code></pre>
<p>The table lists the predicted d13c value for each block (accounting for population and family) and a 95% confidence interval for each predicted value.</p>
<p>Now let’s plot the effects (mean with 95% confidence interval) for each random factor, accounting for all other factors. Again, these are not simply means of observed traits values but marginal effect, accounting for all other factors in the model.</p>
<p>Note: If the plots appear in the ‘Plot’ tab in RStudio, use the arrows to scroll between plots.</p>
<div class="sourceCode" id="cb54"><html><body><pre class="r"><span class="kw pkg">lattice</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/lattice/man/xyplot.html">dotplot</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/lme4/man/ranef.html">ranef</a></span>(<span class="no">mod4</span>,<span class="kw">condVar</span><span class="kw">=</span><span class="fl">TRUE</span>))</pre></body></html></div>
<pre><code>## $family</code></pre>
<p><img src="Week6_vignette_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<pre><code>## 
## $population</code></pre>
<p><img src="Week6_vignette_files/figure-html/unnamed-chunk-25-2.png" width="700"></p>
<p>We can see that there are considerable differences between populations, where ‘blk cyn’ has the highest, and ‘mt watson’ the lowest mean.</p>
<p>We can extract the values with ‘ranef’. Again, they are deviations from the global intercept (block 1). We are now ready to correlate the population effect to the environmental data.</p>
<div class="sourceCode" id="cb57"><html><body><pre class="r"><span class="no">prov.eff</span> <span class="kw">&lt;-</span> <span class="kw pkg">lme4</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/lme4/man/ranef.html">ranef</a></span>(<span class="no">mod4</span>)$<span class="no">population</span>
<span class="no">fam.eff</span> <span class="kw">&lt;-</span> <span class="kw pkg">lme4</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/lme4/man/ranef.html">ranef</a></span>(<span class="no">mod4</span>)$<span class="no">family</span>
<span class="no">prov.eff</span></pre></body></html></div>
<pre><code>##              (Intercept)
## armstrong    -0.16372457
## blk cyn       0.56424812
## echo lk      -0.10480714
## flume         0.37270696
## hvn          -0.25055204
## incline lk    0.21873458
## jakes         0.03776884
## meiss mdw    -0.18557714
## montreal cyn -0.02673127
## mt watson    -0.46206634</code></pre>
<div class="sourceCode" id="cb59"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">fam.eff</span>)</pre></body></html></div>
<pre><code>##     (Intercept)
## 59  0.269660618
## 60 -0.036688490
## 61 -0.007035082
## 63 -0.259031610
## 64  0.087029419
## 65 -0.168188771</code></pre>
</div>
<div id="b-correlation-of-population-effect-with-environment" class="section level3">
<h3 class="hasAnchor">
<a href="#b-correlation-of-population-effect-with-environment" class="anchor"></a>b) Correlation of population effect with environment</h3>
<p>We have population effects estimated for 10 populations, and bioclimatic data for 30 plots (3 plots x 10 populations). If we used all 157 rows of the site data set, this would inflate our sample size. To avoid this, we first extract 30 unique rows (one per plot) from the site data ‘<a href="mailto:WWP.ecogen@E" class="email">WWP.ecogen@E</a>’. Then we use’ match’ to find for each plot the corresponding population. This returns the row number (‘index’) and we use it to extract the corresponding population effect.</p>
<div class="sourceCode" id="cb61"><html><body><pre class="r"><span class="no">Plot.data</span> <span class="kw">&lt;-</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">WWP.ecogen</span>@<span class="kw">E</span>,
                    <span class="kw">population</span><span class="kw">=</span><span class="no">WWP.ecogen</span>@<span class="kw">S</span>$<span class="no">population</span>))
<span class="no">index</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="no">Plot.data</span>$<span class="no">population</span>, <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span>(<span class="no">prov.eff</span>))
<span class="no">dat2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">prov.eff</span><span class="kw">=</span><span class="no">prov.eff</span>[<span class="no">index</span>,<span class="fl">1</span>], <span class="no">Plot.data</span>[,-<span class="fl">8</span>])</pre></body></html></div>
<p>Now we estimate the correlation between the population effect and each bioclimatic variable. We summarize the results in a correlation matrix plot.</p>
<div class="sourceCode" id="cb62"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/pairs.html">pairs</a></span>(<span class="no">dat2</span>, <span class="kw">lower.panel</span><span class="kw">=</span><span class="no">panel.smooth</span>, <span class="kw">upper.panel</span><span class="kw">=</span><span class="no">panel.cor</span>,
      <span class="kw">diag.panel</span><span class="kw">=</span><span class="no">panel.hist</span>)</pre></body></html></div>
<p><img src="Week6_vignette_files/figure-html/unnamed-chunk-28-1.png" width="720"></p>
<p>How to read this plot:</p>
<ul>
<li>Each diagonal cell shows the name and histogram of one variable.</li>
<li>Each cell in the upper triangle shows the correlation coefficient r, and the p-value for the correlation, between the two variables that define the corresponding row and column. (A larger font indicates a stronger correlation, though this seems to be inconsistent).</li>
<li>Each cell in the lower triangle shows a scatterplot of the two variables that define the corresponding row and column. A smooth regression line is shown in red.</li>
</ul>
<p><strong>Questions:</strong></p>
<ul>
<li>Which site variables show the strongest correlation with the population effect?</li>
<li>Which site variables are significantly correlated with the population effect?</li>
<li>OVerall, does the population effect appear to be related to the environment?</li>
</ul>
</div>
</div>
<div id="r-exercise-week-6" class="section level2">
<h2 class="hasAnchor">
<a href="#r-exercise-week-6" class="anchor"></a>7. R Exercise Week 6</h2>
<p><strong>Task:</strong> Test whether observed heterozygosity of <em>Pulsatilla vulgaris</em> adults depends on census population size. Fit a model at the individual level where you include a random effect for population.</p>
<p><strong>Hints:</strong></p>
<ol style="list-style-type: lower-alpha">
<li>
<p><strong>Load packages</strong>:</p>
<ul>
<li>Please install the package <code>inbreedR</code> (to calculate individual measures of heterozygosity) from CRAN, if it is not yet installed.</li>
<li>You may want to load the packages <code>dplyr</code> and <code>ggplot2</code>. Alternatively, you can use <code>::</code> to call functions from packages.</li>
</ul>
</li>
<li>
<p><strong>Import data and extract adults</strong>.</p>
<ul>
<li>Use the code below to import the data.</li>
<li>Use <code><a href="https://dplyr.tidyverse.org/reference/filter.html">dplyr::filter</a></code> to extract adults with <code>OffID == 0</code>.</li>
</ul>
<p><code>Pulsatilla &lt;- read.csv(system.file("extdata","pulsatilla_genotypes.csv",                           package = "LandGenCourse"))</code></p>
</li>
<li>
<p><strong>Calculate multilocus heterozygosity</strong>: Use package <code>inbreedR</code> to calculate multilocus heterozygosity for each adult.</p>
<ul>
<li>Use the function <code>inbreedR::convert_raw(x)</code>, where x is the matrix of genotypes only (no ID or other non-genetic data), with two columns per locus. Check the help file of the function <code>convert_raw</code>.</li>
<li>Use the function <code>inbreedR::MLH</code> to calculate observed heterozygosity for each individual.</li>
<li>Add the result as a variable <code>het</code> to the adults dataset.</li>
</ul>
<p>Example code from <code>inbreedR::MLH</code> help file: <code>data(mouse_msats)   genotypes &lt;- convert_raw(mouse_msats)   het &lt;- MLH(genotypes)</code></p>
</li>
<li>
<p><strong>Add population-level data</strong>:</p>
<ul>
<li>Import the file “pulsatilla_population.csv” with the code below.</li>
<li>Do a left-join to add the data to your adults dataset.</li>
<li>Check the dataset.</li>
</ul>
<p><code>Pop.data &lt;- read.csv(system.file("extdata", "pulsatilla_population.csv",                           package = "LandGenCourse"))</code></p>
</li>
<li><p><strong>Scatterplot with regression line</strong>: Use ggplot2 to create a scatterplot of adult heterozygosity against census population size (<code>population.size</code>), with a regression line.</p></li>
<li><p><strong>Fit linear mixed model</strong>: Adapt code from section 3.c to perform a regression of individual-level observed heterozygosity (response variable) on census population size (predictor), including population as a random effect. Fit the model with REML and print a summary.</p></li>
<li><p><strong>Test fixed effect</strong>: Adapt code from section 2.f to test the fixed effect with function <code><a href="https://rdrr.io/pkg/car/man/Anova.html">car::Anova</a></code>.</p></li>
<li><p><strong>Check residual plots</strong>: Adapt code from section 2.d to create residual plots.</p></li>
</ol>
<p><strong>Questions:</strong> There is one influential point in the regression analysis:</p>
<ul>
<li>What was the direction of the relationship, did heterozygosity increase or decrease with census population size?</li>
<li>Was the fixed effect statistically significant?</li>
<li>Was the model valid, or was there a problem with the residual plots?</li>
<li>What would be the main issue, and what remedy could you suggest?</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://sites.utm.utoronto.ca/wagnerlab/">Helene Wagner</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
